<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>A Robust Tool for Performing Transcriptome Profiling on Cancer Data â€¢ RNAsum</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="A Robust Tool for Performing Transcriptome Profiling on Cancer Data">
<meta property="og:description" content="RNA-seq reporting workflow designed to post-process,
    summarise and visualise an output from bcbio-nextgen RNA-seq or Dragen
    RNA pipelines. Its main application is to complement genome-based
    findings and provide additional evidence for detected alterations.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">RNAsum</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/report_structure.html">RNAsum Sections</a>
    </li>
    <li>
      <a href="articles/tcga_projects_summary.html">TCGA Projects</a>
    </li>
    <li>
      <a href="articles/workflow.html">Workflow Details</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/umccr/RNAsum/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">

<ul>
<li>
<a href="#rnasum" id="toc-rnasum">RNAsum</a>
<ul>
<li><a href="#installation" id="toc-installation">Installation</a></li>
<li><a href="#workflow" id="toc-workflow">Workflow</a></li>
<li>
<a href="#reference-data" id="toc-reference-data">Reference data</a>
<ul>
<li><a href="#external-reference-cohorts" id="toc-external-reference-cohorts">External reference cohorts</a></li>
<li><a href="#internal-reference-cohort" id="toc-internal-reference-cohort">Internal reference cohort</a></li>
</ul>
</li>
<li>
<a href="#input-data" id="toc-input-data">Input data</a>
<ul>
<li><a href="#wts" id="toc-wts">WTS</a></li>
<li><a href="#wgs" id="toc-wgs">WGS</a></li>
</ul>
</li>
<li>
<a href="#usage" id="toc-usage">Usage</a>
<ul>
<li><a href="#run-via-aws-batch" id="toc-run-via-aws-batch">Run via AWS-Batch</a></li>
<li><a href="#examples" id="toc-examples">Examples</a></li>
<li><a href="#output" id="toc-output">Output</a></li>
</ul>
</li>
</ul>
</li>
</ul>
<!-- README.md is generated from README.Rmd. Please edit that file --><div class="section level1">
<div class="page-header"><h1 id="rnasum">RNAsum<a class="anchor" aria-label="anchor" href="#rnasum"></a>
</h1></div>
<p><code>RNAsum</code> is an R package that can post-process, summarise and visualise outputs from <a href="https://sapac.illumina.com/products/by-type/informatics-products/basespace-sequence-hub/apps/edico-genome-inc-dragen-rna-pipeline.html" class="external-link">DRAGEN RNA</a> or <a href="https://bcbio-nextgen.readthedocs.io/en/latest/contents/bulk_rnaseq.html" class="external-link">bcbio-nextgen RNA-seq</a> pipelines. Its main application is to complement genome-based findings from the <a href="https://github.com/umccr/umccrise" class="external-link">umccrise</a> pipeline and to provide additional evidence for detected alterations.</p>
<p><strong>DOCS</strong>: <a href="https://umccr.github.io/rnasum" class="external-link uri">https://umccr.github.io/rnasum</a></p>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<ul>
<li>
<strong>R</strong> package can be installed directly from the <a href="https://github.com/umccr/RNAsum" class="external-link">GitHub source</a>:</li>
</ul>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"umccr/RNAsum"</span><span class="op">)</span> <span class="co"># latest master commit</span>
<span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"umccr/RNAsum@v0.0.X"</span><span class="op">)</span> <span class="co"># version 0.0.X</span>
<span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"umccr/RNAsum@abcde"</span><span class="op">)</span> <span class="co"># commit abcde</span>
<span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"umccr/RNAsum#123"</span><span class="op">)</span> <span class="co"># PR 123</span></code></pre></div>
<ul>
<li>
<strong>Conda</strong> package is available from the Anaconda <a href="https://anaconda.org/umccr/r-rnasum" class="external-link">umccr channel</a>:</li>
</ul>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> install r-rnasum==0.0.X <span class="at">-c</span> umccr <span class="at">-c</span> conda-forge <span class="at">-c</span> bioconda</span></code></pre></div>
<ul>
<li>
<strong>Docker</strong> image is available from the <a href="https://github.com/umccr/RNAsum/pkgs/container/rnasum" class="external-link">GitHub Container Registy</a>:</li>
</ul>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="ex">docker</span> pull ghcr.io/umccr/rnasum:latest</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="workflow">Workflow<a class="anchor" aria-label="anchor" href="#workflow"></a>
</h2>
<p>The pipeline consists of five main components illustrated and briefly described below. For more details, see <a href="/workflow.html">workflow.md</a>.</p>
<p><img src="articles/img/RNAsum_workflow.png" width="100%"></p>
<ol style="list-style-type: decimal">
<li><p>Collect patient <strong>WTS data</strong> from the <code>DRAGEN RNA</code> or <code>bcbio-nextgen RNA-seq</code> pipeline including per-gene read counts and gene fusions.</p></li>
<li><p>Add expression data from <strong><a href="#reference-data">reference cohorts</a></strong> to get an idea about the expression levels of genes of interest in other cancer patient cohorts. The read counts are normalised, transformed and converted into a scale that allows to present the patientâ€™s expression measurements in the context of the reference cohorts.</p></li>
<li><p>Supply <strong>genome-based findings</strong> from whole-genome sequencing (WGS) data to focus on genes of interest and to provide additional evidence for dysregulation of mutated genes, or genes located within detected structural variants (SVs) or copy-number (CN) altered regions. <code>RNAsum</code> is designed to be compatible with WGS patient outputs generated from <code>umccrise</code>.</p></li>
<li><p>Collate results with knowledge derived from in-house resources and public databases to provide additional sources of evidence for clinical significance of altered genes e.g.Â to flag variants with clinical significance or potential druggable targets.</p></li>
<li><p>The final product is an interactive HTML report with searchable tables and plots presenting expression levels of the genes of interest. The report consists of several sections described <a href="./report_structure.html">here</a>.</p></li>
</ol>
</div>
<div class="section level2">
<h2 id="reference-data">Reference data<a class="anchor" aria-label="anchor" href="#reference-data"></a>
</h2>
<p>The reference expression data are available for <strong>33 cancer types</strong> and were derived from <a href="#external-reference-cohorts">external</a> (<a href="https://tcga-data.nci.nih.gov/" class="external-link">TCGA</a>) and <a href="#internal-reference-cohort">internal</a> (<a href="https://research.unimelb.edu.au/centre-for-cancer-research/our-research/precision-oncology-research-group" class="external-link">UMCCR</a>) resources.</p>
<div class="section level3">
<h3 id="external-reference-cohorts">External reference cohorts<a class="anchor" aria-label="anchor" href="#external-reference-cohorts"></a>
</h3>
<p>In order to explore expression changes in the patient, we have built a high-quality pancreatic cancer reference cohort.</p>
<p>Depending on the tissue from which the patientâ€™s sample was taken, one of <strong>33 cancer datasets</strong> from TCGA can be used as a reference cohort for comparing expression changes in genes of interest of the patient. Additionally, 10 samples from each of the 33 TCGA datasets were combined to create the <strong><a href="./TCGA_projects_summary.md#pan-cancer-dataset">Pan-Cancer dataset</a></strong>, and for some cohorts <strong><a href="./TCGA_projects_summary.md#extended-datasets">extended sets</a></strong> are also available. All available datasets are listed in <strong><a href="./TCGA_projects_summary.html">TCGA projects summary table</a></strong>. These datasets have been processed using methods described in the <a href="https://github.com/umccr/TCGA-data-harmonization/blob/master/expression/README.md#gdc-counts-data" class="external-link">TCGA-data-harmonization</a> repository. The dataset of interest can be specified by using one of the TCGA project IDs for the <code>RNAsum</code> <code>--dataset</code> argument (see <a href="./README.md#arguments">Arguments</a>).</p>
</div>
<div class="section level3">
<h3 id="internal-reference-cohort">Internal reference cohort<a class="anchor" aria-label="anchor" href="#internal-reference-cohort"></a>
</h3>
<p>The publicly available TCGA datasets are expected to demonstrate prominent <a href="https://www.ncbi.nlm.nih.gov/pubmed/20838408" class="external-link">batch effects</a> when compared to the in-house WTS data due to differences in applied experimental procedures and analytical pipelines. Moreover, TCGA data may include samples from tissue material of lower quality and cellularity compared to samples processed using local protocols. To address these issues, we have built a high-quality internal reference cohort processed using the same pipelines as input data (see <a href="./workflow.md#data-pre-processing">data pre-processing</a>).</p>
<p>This internal reference set of <strong>40 pancreatic cancer samples</strong> is based on WTS data generated at <strong><a href="https://research.unimelb.edu.au/centre-for-cancer-research/our-research/precision-oncology-research-group" class="external-link">UMCCR</a></strong> and processed with the <strong>bcbio-nextgen RNA-seq</strong> pipeline to minimise potential batch effects between investigated samples and the reference cohort and to make sure the data are comparable. The internal reference cohort assembly is summarised in the <a href="https://github.com/umccr/Pancreatic-data-harmonization/tree/master/expression/in-house" class="external-link">Pancreatic-data-harmonization</a> repository.</p>
<p><strong>Note</strong></p>
<p>The are two rationales for using the internal reference cohort:</p>
<ol style="list-style-type: decimal">
<li>In case of <strong>pancreatic cancer samples</strong> this cohort is used:
<ul>
<li>in <strong><em>batch effects correction</em></strong>
</li>
<li>as a reference point for <strong><em>comparing per-gene expression levels</em></strong> observed in the data of the patient of interest and data from other pancreatic cancer patients.</li>
</ul>
</li>
<li>In case of samples from <strong>any cancer type</strong> the data from the internal reference cohort is used in the <strong><em>batch effects correction</em></strong> procedure performed to minimise technical-related variation in the data.</li>
</ol>
</div>
</div>
<div class="section level2">
<h2 id="input-data">Input data<a class="anchor" aria-label="anchor" href="#input-data"></a>
</h2>
<p><code>RNAsum</code> accepts <a href="#wts">WTS</a> data processed by the <code>DRAGEN RNA</code> or <code>bcbio-nextgen RNA-seq</code> pipeline. Additionally, the WTS data can be integrated with <a href="#wgs">WGS</a>-based data processed using the <code>umccrise</code> pipeline. In the latter case, the genome-based findings from the corresponding patient sample are incorporated into the report and are used as a primary source for expression profile prioritisation.</p>
<div class="section level3">
<h3 id="wts">WTS<a class="anchor" aria-label="anchor" href="#wts"></a>
</h3>
<p>The only required WTS input data are <strong>read counts</strong> provided in a quantification file from either the <code>DRAGEN RNA</code> or <code>bcbio-nextgen RNA-seq</code> pipeline.</p>
<div class="section level4">
<h4 id="dragen-rna">DRAGEN RNA<a class="anchor" aria-label="anchor" href="#dragen-rna"></a>
</h4>
<p>The table below lists all input data accepted in <code>RNAsum</code>:</p>
<table class="table">
<colgroup>
<col width="12%">
<col width="49%">
<col width="34%">
<col width="3%">
</colgroup>
<thead><tr class="header">
<th>Input file</th>
<th>Tool</th>
<th>Example</th>
<th>Required</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Quantified transcript <strong>abundances</strong>
</td>
<td>
<a href="https://salmon.readthedocs.io/en/latest/salmon.html" class="external-link">salmon</a> (<a href="https://salmon.readthedocs.io/en/latest/file_formats.html#fileformats" class="external-link">description</a>)</td>
<td><a href="/inst/rawdata/test_data/dragen/TEST.quant.sf">TEST.quant.sf</a></td>
<td><strong>Yes</strong></td>
</tr>
<tr class="even">
<td>
<strong>Fusion gene</strong> list</td>
<td><a href="https://sapac.illumina.com/products/by-type/informatics-products/basespace-sequence-hub/apps/edico-genome-inc-dragen-rna-pipeline.html" class="external-link">DRAGEN RNA</a></td>
<td><a href="/inst/rawdata/test_data/dragen/test_sample_WTS.fusion_candidates.final">TEST.fusion_candidates.final</a></td>
<td>No</td>
</tr>
</tbody>
</table>
<p>These files are expected to be organised in the following structure:</p>
<pre class="text"><code>|
|____&lt;SampleName&gt;
  |____&lt;SampleName&gt;quant.sf
  |____&lt;SampleName&gt;.fusion_candidates.final</code></pre>
</div>
<div class="section level4">
<h4 id="bcbio-nextgen-legacy">bcbio-nextgen (legacy)<a class="anchor" aria-label="anchor" href="#bcbio-nextgen-legacy"></a>
</h4>
<p>The table below lists all input data accepted in <code>RNAsum</code>:</p>
<table class="table">
<colgroup>
<col width="10%">
<col width="35%">
<col width="51%">
<col width="2%">
</colgroup>
<thead><tr class="header">
<th>Input file</th>
<th>Tool</th>
<th>Example</th>
<th>Required</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Quantified transcript <strong>abundances</strong>
</td>
<td>
<a href="https://pachterlab.github.io/kallisto/about" class="external-link">kallisto</a> (<a href="https://pachterlab.github.io/kallisto/starting#results" class="external-link">description</a>)</td>
<td>
<a href="/inst/rawdata/test_data/final/test_sample_WTS/kallisto/abundance.tsv">abundance.tsv</a>, <a href="https://www.rna-seqblog.com/rpkm-fpkm-and-tpm-clearly-explained/" class="external-link">Transcripts Per Million</a>
</td>
<td><strong>Yes</strong></td>
</tr>
<tr class="even">
<td>
<strong>Fusion gene</strong> list</td>
<td>
<a href="https://arriba.readthedocs.io/en/latest" class="external-link">arriba</a>, <a href="https://github.com/pmelsted/pizzly" class="external-link">pizzly</a>
</td>
<td>
<a href="/inst/rawdata/test_data/final/test_sample_WTS/arriba/fusions.tsv">fusions.tsv</a>, <a href="/inst/rawdata/test_data/final/test_sample_WTS/pizzly/test_sample_WTS-flat.tsv">test_sample_WTS-flat.tsv</a>
</td>
<td>No</td>
</tr>
<tr class="odd">
<td>
<strong>Fusion gene</strong> plots</td>
<td><a href="https://arriba.readthedocs.io/en/latest/visualization/" class="external-link">arriba plot</a></td>
<td><a href="/inst/rawdata/test_data/final/test_sample_WTS/arriba/fusions.pdf">fusions.pdf</a></td>
<td>No</td>
</tr>
</tbody>
</table>
<p>These files are expected to be organised in the following structure:</p>
<pre class="text"><code>|
|____&lt;SampleName&gt;
  |____kallisto
  | |____abundance.tsv
  |____pizzly
  | |____&lt;SampleName&gt;-flat.tsv
  |____arriba
    |____fusions.pdf
    |____fusions.tsv</code></pre>
<p><strong>Note</strong></p>
<p><code>pizzly</code> outputs two fusion gene files by default: - <em>&lt;sample_name&gt;-flat.tsv</em> lists all (<strong>unfiltered</strong>) the gene fusions - <em>&lt;sample_name&gt;-flat-filtered.tsv</em> lists <strong>filtered</strong> gene fusions</p>
<p><code>RNAsum</code> makes use of gene fusions listed in the <strong>unfiltered</strong> file since it was noted that some genuine fusions (based on WGS data and curation efforts) get filtered out.</p>
</div>
</div>
<div class="section level3">
<h3 id="wgs">WGS<a class="anchor" aria-label="anchor" href="#wgs"></a>
</h3>
<p><code>RNAsum</code> is designed to be compatible with WGS outputs generated from <code>umccrise</code>.</p>
<p>The table below lists all input data accepted in <code>RNAsum</code>:</p>
<table class="table">
<colgroup>
<col width="7%">
<col width="32%">
<col width="55%">
<col width="4%">
</colgroup>
<thead><tr class="header">
<th>Input file</th>
<th>Tool</th>
<th>Example</th>
<th>Required</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><strong>SNVs/Indels</strong></td>
<td><a href="https://github.com/sigven/pcgr" class="external-link">PCGR</a></td>
<td><a href="/inst/rawdata/test_data/umccrised/test_sample_WGS/small_variants/pcgr.snvs_indels.tiers.tsv">pcgr.snvs_indels.tiers.tsv</a></td>
<td>No</td>
</tr>
<tr class="even">
<td><strong>CNVs</strong></td>
<td><a href="https://github.com/hartwigmedical/hmftools/tree/master/purple" class="external-link">PURPLE</a></td>
<td><a href="/inst/rawdata/test_data/umccrised/test_sample_WGS/purple/purple.gene.cnv">purple.cnv.gene.tsv</a></td>
<td>No</td>
</tr>
<tr class="odd">
<td><strong>SVs</strong></td>
<td><a href="https://github.com/Illumina/manta" class="external-link">Manta</a></td>
<td><a href="/inst/rawdata/test_data/umccrised/test_sample_WGS/structural/sv-prioritize-manta.tsv">sv-prioritize-manta.tsv</a></td>
<td>No</td>
</tr>
</tbody>
</table>
<p>These files are expected to be organised in the following structure:</p>
<pre class="text"><code>|
|____umccrised
  |____&lt;SampleName&gt;
    |____pcgr
    | |____&lt;SampleName&gt;-somatic.pcgr.snvs_indels.tiers.tsv
    |____purple
    | |____&lt;SampleName&gt;.purple.gene.cnv
    |____structural
      |____&lt;SampleName&gt;-manta.tsv</code></pre>
</div>
</div>
<div class="section level2">
<h2 id="usage">Usage<a class="anchor" aria-label="anchor" href="#usage"></a>
</h2>
<div class="sourceCode" id="cb7"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="va">rnasum_cli</span><span class="op">=</span><span class="va">$(</span><span class="ex">Rscript</span> <span class="at">-e</span> <span class="st">'x = system.file("cli", package = "RNAsum"); cat(x, "\n")'</span> <span class="kw">|</span> <span class="fu">xargs</span><span class="va">)</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="bu">export</span> <span class="va">PATH</span><span class="op">=</span><span class="st">"</span><span class="va">${rnasum_cli}</span><span class="st">:</span><span class="va">${PATH}</span><span class="st">"</span></span></code></pre></div>
<div class="sourceCode" id="cb8"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span> rnasum.R <span class="sc">--</span>version</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>rnasum.R x.x.x</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span> rnasum.R <span class="sc">--</span>help</span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>bash<span class="sc">:</span> line <span class="dv">8</span><span class="sc">:</span> rnasum.R<span class="sc">:</span> command not found</span></code></pre></div>
<p><strong>Note</strong></p>
<p>Human reference genome <strong><em><a href="https://www.ncbi.nlm.nih.gov/assembly/GCF_000001405.39" class="external-link">GRCh38</a></em></strong> (<em>Ensembl</em> based annotation version <strong><em>86</em></strong>) is used for gene annotation by default. Alternatively, human reference genome <a href="https://www.ncbi.nlm.nih.gov/assembly/GCF_000001405.13/" class="external-link">GRCh37</a> (<em>Ensembl</em> based annotation version <em>75</em>) is used when argument <code>grch_version</code> is set to <code>37</code>.</p>
<div class="section level3">
<h3 id="run-via-aws-batch">Run via AWS-Batch<a class="anchor" aria-label="anchor" href="#run-via-aws-batch"></a>
</h3>
<p>At UMCCR we run <code>RNAsum</code> in production via <a href="https://aws.amazon.com/batch/" class="external-link">AWS-batch</a>. Before running, collate the following information:</p>
<ul>
<li>S3 path to umccrise results.</li>
<li>S3 path to WTS data results (bcbio).</li>
<li>
<a href="https://github.com/umccr/RNAsum/blob/master/TCGA_projects_summary.md" class="external-link">Reference dataset</a> to use for the sample.</li>
</ul>
<p>Next, follow these steps:</p>
<ol style="list-style-type: decimal">
<li>Configure the AWS CLI to access data on S3 (for the UMCCR account see our <a href="https://github.com/umccr/wiki/blob/master/computing/cloud/amazon/aws_cli.md#configuration" class="external-link">wiki</a>).</li>
<li>Run the <code>aws lambda invoke</code> command, based on the example below:</li>
</ol>
<div class="sourceCode" id="cb9"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="ex">aws</span> lambda invoke <span class="dt">\</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">--region</span> ap-southeast-2 <span class="dt">\</span></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">--function-name</span> wts_report_trigger_lambda_prod <span class="dt">\</span></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">--cli-binary-format</span> raw-in-base64-out <span class="dt">\</span></span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">--payload</span> <span class="st">'{"dataDirWGS":"Scott-SFRC/SBJ0000/WGS/2021-11-24/umccrised/SBJ0000__SB0000_something", "dataDirWTS":"Scott-SFRC/SBJ0000/WTS/2021-11-07/final/SBJ0000_something", "refDataset":"UCEC"}'</span> <span class="dt">\</span></span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>  /tmp/lambda.output</span></code></pre></div>
<p>The <code>RNAsum</code> HTML report for this sample will be created in the <code>umccr-primary-data-prod</code> S3 bucket, under <code>Scott-SFRC/SBJ0000/WTS/2021-11-07/RNAsum</code>. To use primary data from a different S3 bucket, use <code>"dataBucket":"a-different-bucket"</code> in the <code>aws lambda invoke</code> command. Also set <code>resultBucket</code> accordingly.</p>
</div>
<div class="section level3">
<h3 id="examples">Examples<a class="anchor" aria-label="anchor" href="#examples"></a>
</h3>
<p>Below are <code>RNAsum</code> CLI commands for generating HTML reports under different data availability scenarios:</p>
<ol style="list-style-type: decimal">
<li><a href="#id_1-wts-data-only">WTS data only</a></li>
<li><a href="#id_2-wts-and-wgs-data">WTS and WGS data</a></li>
<li><a href="#id_3-wts-wgs-and-clinical-data">WTS WGS and clinical data</a></li>
</ol>
<p><strong>Note</strong></p>
<ul>
<li>Example data is provided in the <code>/inst/rawdata/test_data</code> folder of the GitHub <a href="https://github.com/umccr/RNAsum" class="external-link">repo</a>.</li>
<li>The <code>RNAsum</code> runtime should be less than <strong>20 minutes</strong> using <strong>16GB RAM</strong> memory and <strong>1 CPU</strong>.</li>
</ul>
<div class="section level4">
<h4 id="id_1-wts-data-only">1. WTS data only<a class="anchor" aria-label="anchor" href="#id_1-wts-data-only"></a>
</h4>
<p>In this scenario, only <a href="#wts">WTS</a> data will be used and only expression levels of key <strong><a href="https://github.com/umccr/umccrise/blob/master/workflow.md#key-cancer-genes" class="external-link"><code>Cancer genes</code></a></strong>, <strong><code>Fusion genes</code></strong>, <strong><code>Immune markers</code></strong> and homologous recombination deficiency genes (<strong><code>HRD genes</code></strong>) will be reported. Moreover, gene fusions reported in the <code>Fusion genes</code> report section will not contain information about evidence from genome-based data. A subset of the TCGA pancreatic adenocarcinoma dataset is used as the reference cohort (<code>--dataset TEST</code>).</p>
<div class="section level5">
<h5 id="dragen-rna-1">DRAGEN RNA<a class="anchor" aria-label="anchor" href="#dragen-rna-1"></a>
</h5>
<div class="sourceCode" id="cb10"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="ex">rnasum.R</span> <span class="dt">\</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">--sample_name</span> test_sample_WTS <span class="dt">\</span></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">--dataset</span> TEST <span class="dt">\</span></span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">--dragen_rnaseq</span> <span class="va">$(</span><span class="bu">pwd</span><span class="va">)</span>/../rawdata/test_data/dragen <span class="dt">\</span></span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">--report_dir</span> <span class="va">$(</span><span class="bu">pwd</span><span class="va">)</span>/../rawdata/test_data/dragen/RNAsum <span class="dt">\</span></span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">--save_tables</span> FALSE</span></code></pre></div>
<p>The HTML report <code>test_sample_WTS.RNAsum.html</code> will be created in the <code>../rawdata/test_data/dragen/RNAsum</code> folder.</p>
</div>
<div class="section level5">
<h5 id="bcbio-nextgen-legacy-1">bcbio-nextgen (legacy)<a class="anchor" aria-label="anchor" href="#bcbio-nextgen-legacy-1"></a>
</h5>
<div class="sourceCode" id="cb11"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="ex">rnasum.R</span> <span class="dt">\</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">--sample_name</span> test_sample_WTS <span class="dt">\</span></span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">--dataset</span> TEST <span class="dt">\</span></span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">--bcbio_rnaseq</span> <span class="va">$(</span><span class="bu">pwd</span><span class="va">)</span>/../rawdata/test_data/final/test_sample_WTS <span class="dt">\</span></span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">--report_dir</span> <span class="va">$(</span><span class="bu">pwd</span><span class="va">)</span>/../rawdata/test_data/final/test_sample_WTS/RNAsum <span class="dt">\</span></span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">--save_tables</span> FALSE</span></code></pre></div>
<p>The HTML report <code>test_sample_WTS.RNAsum.html</code> will be created in the <code>../rawdata/test_data/final/test_sample_WTS/RNAsum</code> folder.</p>
</div>
</div>
<div class="section level4">
<h4 id="id_2-wts-and-wgs-data">2. WTS and WGS data<a class="anchor" aria-label="anchor" href="#id_2-wts-and-wgs-data"></a>
</h4>
<p>This is the <strong>most frequent and preferred case</strong>, in which the <a href="#wgs">WGS</a>-based findings will be used as a primary source for expression profile prioritisation. The genome-based results can be incorporated into the report by specifying the location of the corresponding <code>umccrise</code> output files (including results from <code>PCGR</code>, <code>PURPLE</code>, and <code>Manta</code>) using the <code>--umccrise</code> argument. The <strong><code>Mutated genes</code></strong>, <strong><code>Structural variants</code></strong> and <strong><code>CN altered genes</code></strong> report sections will contain information about expression levels of the mutated genes, genes located within detected SVs and CN altered regions, respectively. The results in the <strong><code>Fusion genes</code></strong> section will be ordered based on the evidence from genome-based data. A subset of the TCGA pancreatic adenocarcinoma dataset is used as reference cohort (<code>--dataset TEST</code>).</p>
<div class="section level5">
<h5 id="dragen-rna-2">DRAGEN RNA<a class="anchor" aria-label="anchor" href="#dragen-rna-2"></a>
</h5>
<div class="sourceCode" id="cb12"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="ex">rnasum.R</span> <span class="dt">\</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">--sample_name</span> test_sample_WTS <span class="dt">\</span></span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">--dataset</span> TEST <span class="dt">\</span></span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">--dragen_rnaseq</span> <span class="va">$(</span><span class="bu">pwd</span><span class="va">)</span>/../rawdata/test_data/dragen <span class="dt">\</span></span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">--report_dir</span> <span class="va">$(</span><span class="bu">pwd</span><span class="va">)</span>/../rawdata/test_data/dragen/RNAsum <span class="dt">\</span></span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">--umccrise</span> <span class="va">$(</span><span class="bu">pwd</span><span class="va">)</span>/../rawdata/test_data/umccrised/test_sample_WGS <span class="dt">\</span></span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">--save_tables</span> FALSE</span></code></pre></div>
<p>The HTML report <code>test_sample_WTS.RNAsum.html</code> will be created in the <code>../rawdata/test_data/dragen/RNAsum</code> folder.</p>
</div>
<div class="section level5">
<h5 id="bcbio-nextgen-legacy-2">bcbio-nextgen (legacy)<a class="anchor" aria-label="anchor" href="#bcbio-nextgen-legacy-2"></a>
</h5>
<div class="sourceCode" id="cb13"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="ex">rnasum.R</span> <span class="dt">\</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">--sample_name</span> test_sample_WTS <span class="dt">\</span></span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">--dataset</span> TEST <span class="dt">\</span></span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">--bcbio_rnaseq</span> <span class="va">$(</span><span class="bu">pwd</span><span class="va">)</span>/../rawdata/test_data/final/test_sample_WTS <span class="dt">\</span></span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">--report_dir</span> <span class="va">$(</span><span class="bu">pwd</span><span class="va">)</span>/../rawdata/test_data/final/test_sample_WTS/RNAsum <span class="dt">\</span></span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">--umccrise</span> <span class="va">$(</span><span class="bu">pwd</span><span class="va">)</span>/../rawdata/test_data/umccrised/test_sample_WGS <span class="dt">\</span></span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">--save_tables</span> FALSE</span></code></pre></div>
<p>The HTML report <code>test_sample_WTS.RNAsum.html</code> will be created in the <code>../rawdata/test_data/final/test_sample_WTS/RNAsum</code> folder.</p>
</div>
</div>
<div class="section level4">
<h4 id="id_3-wts-wgs-and-clinical-data">3. WTS WGS and clinical data<a class="anchor" aria-label="anchor" href="#id_3-wts-wgs-and-clinical-data"></a>
</h4>
<p>For samples derived from subjects, for which clinical information is available, a treatment regimen timeline can be added to the HTML report. This can be added by specifying the location of a relevant excel spreadsheet (see example <a href="./data/test_data/test_clinical_data.xlsx">test_clinical_data.xlsx</a>) using the <code>--clinical_info</code> argument. In this spreadsheet, at least one of the following columns is expected: <code>NEOADJUVANT REGIMEN</code>, <code>ADJUVANT REGIMEN</code>, <code>FIRST LINE REGIMEN</code>, <code>SECOND LINE REGIMEN</code> or <code>THIRD LINE REGIMEN</code>, along with <code>START</code> and <code>STOP</code> dates of corresponding treatments. A subset of the TCGA pancreatic adenocarcinoma dataset is used as the reference cohort (<code>--dataset TEST</code>).</p>
<div class="section level5">
<h5 id="dragen-rna-3">DRAGEN RNA<a class="anchor" aria-label="anchor" href="#dragen-rna-3"></a>
</h5>
<div class="sourceCode" id="cb14"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="ex">rnasum.R</span> <span class="dt">\</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">--sample_name</span> test_sample_WTS <span class="dt">\</span></span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">--dataset</span> TEST <span class="dt">\</span></span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">--dragen_rnaseq</span> <span class="va">$(</span><span class="bu">pwd</span><span class="va">)</span>/../rawdata/test_data/dragen <span class="dt">\</span></span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">--report_dir</span> <span class="va">$(</span><span class="bu">pwd</span><span class="va">)</span>/../rawdata/test_data/dragen/RNAsum <span class="dt">\</span></span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">--umccrise</span> <span class="va">$(</span><span class="bu">pwd</span><span class="va">)</span>/../rawdata/test_data/umccrised/test_sample_WGS <span class="dt">\</span></span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">--save_tables</span> FALSE <span class="dt">\</span></span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">--clinical_info</span> <span class="va">$(</span><span class="bu">pwd</span><span class="va">)</span>/../rawdata/test_data/test_clinical_data.xlsx <span class="dt">\</span></span>
<span id="cb14-9"><a href="#cb14-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">--save_tables</span> FALSE</span></code></pre></div>
<p>The HTML report <code>test_sample_WTS.RNAsum.html</code> will be created in the <code>../rawdata/test_data/stratus/test_sample_WTS_dragen_v3.9.3/RNAsum</code> folder.</p>
</div>
<div class="section level5">
<h5 id="bcbio-nextgen-legacy-3">bcbio-nextgen (legacy)<a class="anchor" aria-label="anchor" href="#bcbio-nextgen-legacy-3"></a>
</h5>
<div class="sourceCode" id="cb15"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="ex">rnasum.R</span> <span class="dt">\</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">--sample_name</span> test_sample_WTS <span class="dt">\</span></span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">--dataset</span> TEST <span class="dt">\</span></span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">--bcbio_rnaseq</span> <span class="va">$(</span><span class="bu">pwd</span><span class="va">)</span>/../rawdata/test_data/final/test_sample_WTS <span class="dt">\</span></span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">--report_dir</span> <span class="va">$(</span><span class="bu">pwd</span><span class="va">)</span>/../rawdata/test_data/final/test_sample_WTS/RNAsum <span class="dt">\</span></span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">--umccrise</span> <span class="va">$(</span><span class="bu">pwd</span><span class="va">)</span>/../rawdata/test_data/umccrised/test_sample_WGS <span class="dt">\</span></span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">--clinical_info</span> <span class="va">$(</span><span class="bu">pwd</span><span class="va">)</span>/../rawdata/test_data/test_clinical_data.xlsx <span class="dt">\</span></span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">--save_tables</span> FALSE</span></code></pre></div>
<p>The HTML report <code>test_sample_WTS.RNAsum.html</code> will be created in the <code>../rawdata/test_data/final/test_sample_WTS/RNAsum</code> folder.</p>
</div>
</div>
</div>
<div class="section level3">
<h3 id="output">Output<a class="anchor" aria-label="anchor" href="#output"></a>
</h3>
<p>The pipeline generates a HTML <strong><em>Patient Transcriptome Summary</em></strong> <strong><a href="#report">report</a></strong> and a <a href="#results">results</a> folder:</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="sc">|</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a><span class="er">|</span>____<span class="sc">&lt;</span>output<span class="sc">&gt;</span></span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>  <span class="er">|</span>____<span class="sc">&lt;</span>SampleName<span class="sc">&gt;</span>.<span class="sc">&lt;</span>output<span class="sc">&gt;</span>.html</span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>____results</span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a>    <span class="sc">|</span>____exprTables</span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a>    <span class="sc">|</span>____glanceExprPlots</span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a>    <span class="sc">|</span>____...</span></code></pre></div>
<div class="section level4">
<h4 id="report">Report<a class="anchor" aria-label="anchor" href="#report"></a>
</h4>
<p>The generated HTML report includes searchable tables and interactive plots presenting expression levels of altered genes, as well as links to public resources describing the genes of interest. The report consists of several sections, including:</p>
<ul>
<li>Input data</li>
<li>Clinical information*</li>
<li>Findings summary</li>
<li>Mutated genes**</li>
<li>Fusion genes</li>
<li>Structural variants**</li>
<li>CN altered genes**</li>
<li>Immune markers</li>
<li>HRD genes</li>
<li>Cancer genes</li>
<li>Drug matching</li>
</ul>
<p>* if clinical information is available; see <code>--clinical_info</code> argument <br> ** if genome-based results are available; see <code>--umccrise</code> argument</p>
<p>Detailed description of the report structure, including result prioritisation and visualisation is available <a href="report_structure.html">here</a>.</p>
</div>
<div class="section level4">
<h4 id="results">Results<a class="anchor" aria-label="anchor" href="#results"></a>
</h4>
<p>The <code>results</code> folder contains intermediate files, including plots and tables that are presented in the HTML report.</p>
</div>
</div>
</div>
</div>

  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/umccr/RNAsum/" class="external-link">Browse source code</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing RNAsum</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Sehrish Kanwal <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0002-5044-4692" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Jacek Marzec <br><small class="roles"> Author </small>  </li>
</ul>
</div>



  </div>
</div>


      <footer><div class="copyright">
  <p></p>
<p>Developed by Sehrish Kanwal, Jacek Marzec.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.5.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
