<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="RNA-seq reporting workflow designed to post-process,
    summarise and visualise an output from bcbio-nextgen RNA-seq or Dragen
    RNA pipelines. Its main application is to complement genome-based
    findings and provide additional evidence for detected alterations.">
<title>A Robust Tool for Performing Transcriptome Profiling on Cancer Data • RNAsum</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="A Robust Tool for Performing Transcriptome Profiling on Cancer Data">
<meta property="og:description" content="RNA-seq reporting workflow designed to post-process,
    summarise and visualise an output from bcbio-nextgen RNA-seq or Dragen
    RNA pipelines. Its main application is to complement genome-based
    findings and provide additional evidence for detected alterations.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="index.html">RNAsum</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.0.2</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="articles/CHANGELOG.html">Changelog</a>
    <a class="dropdown-item" href="articles/development.html">Development</a>
    <a class="dropdown-item" href="articles/report_structure.html">RNAsum Sections</a>
    <a class="dropdown-item" href="articles/tcga_projects_summary.html">TCGA Projects</a>
    <a class="dropdown-item" href="articles/workflow.html">Workflow Details</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/umccr/RNAsum/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><ul>
<li>
<a href="#rnasum">RNAsum</a>
<ul>
<li>
<a href="#installation">Installation</a>
<ul>
<li><a href="#id_1-system-dependencies-prerequisites">1. System Dependencies (Prerequisites)</a></li>
<li><a href="#id_2-install-r-package">2. Install R Package</a></li>
<li><a href="#alternative-installation-methods">Alternative Installation Methods</a></li>
</ul>
</li>
<li><a href="#example-reports">Example reports</a></li>
<li><a href="#workflow">Workflow</a></li>
<li>
<a href="#reference-data">Reference data</a>
<ul>
<li><a href="#external-reference-cohorts">External reference cohorts</a></li>
<li><a href="#internal-reference-cohort">Internal reference cohort</a></li>
</ul>
</li>
<li>
<a href="#usage">Usage</a>
<ul>
<li><a href="#examples">Examples</a></li>
<li><a href="#output">Output</a></li>
</ul>
</li>
</ul>
</li>
</ul>
<!-- README.md is generated from README.Rmd. Please edit that file --><div class="section level1">
<div class="page-header"><h1 id="rnasum">RNAsum<a class="anchor" aria-label="anchor" href="#rnasum"></a>
</h1></div>
<p><code>RNAsum</code> is an R package for integrating and interpreting whole-genome sequencing (WGS) and whole-transcriptome sequencing (WTS) data from individual cancer patient samples.</p>
<p><strong>DOCS</strong>: <a href="https://umccr.github.io/RNAsum" class="external-link uri">https://umccr.github.io/RNAsum</a></p>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>RNAsum can be installed using one of the following three methods.</p>
<div class="section level3">
<h3 id="a-install-from-github">A. Install from GitHub<a class="anchor" aria-label="anchor" href="#a-install-from-github"></a>
</h3>
<ul>
<li>
<strong>R</strong> package can be installed directly from the <a href="https://github.com/umccr/RNAsum" class="external-link">GitHub source</a>:</li>
</ul>
<div class="section level4">
<h4 id="id_1-system-dependencies-prerequisites">1. System Dependencies (Prerequisites)<a class="anchor" aria-label="anchor" href="#id_1-system-dependencies-prerequisites"></a>
</h4>
<p><code>RNAsum</code> depends on <code>pdftools</code>, which requires system-level libraries (poppler, cairo, etc.) to be installed before installing the R package.</p>
<p><strong>For Linux (Ubuntu/Debian):</strong></p>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sudo</span> apt-get install libpoppler-cpp-dev libharfbuzz-dev libfribidi-dev libfreetype6-dev libcairo2-dev libpango1.0-dev</span></code></pre></div>
<p><strong>For macOS:</strong></p>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="ex">brew</span> install poppler</span></code></pre></div>
<p><strong>For Conda / HPC Environments</strong> If you do not have root access (e.g., on a cluster), creating a fresh Conda environment is the most reliable way to provide necessary system libraries:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a dedicated environment with R and system libraries</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> create <span class="at">-n</span> rnasum_env <span class="at">-c</span> conda-forge <span class="at">-c</span> bioconda <span class="dt">\</span></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>  r-base=4.1 <span class="dt">\</span></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>  poppler harfbuzz fribidi freetype pkg-config cairo openssl pango make gxx_linux-64</span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Activate the environment</span></span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> activate rnasum_env</span></code></pre></div>
</div>
<div class="section level4">
<h4 id="id_2-install-r-package">2. Install R Package<a class="anchor" aria-label="anchor" href="#id_2-install-r-package"></a>
</h4>
<p>Once system dependencies are met, you can install the package directly from GitHub.</p>
<p><strong>Note</strong>: The reference data package (RNAsum.data) is large. It is recommended to increase the download timeout limit before installation. Launch R:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="ex">R</span></span></code></pre></div>
<p>Then run the following inside the R console:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># 1. Increase timeout to prevent download failure for RNAsum.data</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span>timeout <span class="op">=</span> <span class="fl">600</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 2. Install via remotes</span></span>
<span><span class="co"># (This will automatically resolve CRAN, Bioconductor, and GitHub dependencies)</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://remotes.r-lib.org" class="external-link">"remotes"</a></span><span class="op">)</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"remotes"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">remotes</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"umccr/RNAsum"</span><span class="op">)</span> <span class="co"># latest main commit</span></span>
<span><span class="fu">remotes</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"umccr/RNAsum@v0.0.X"</span><span class="op">)</span> <span class="co"># version 0.0.X</span></span>
<span><span class="fu">remotes</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"umccr/RNAsum@abcde"</span><span class="op">)</span> <span class="co"># commit abcde</span></span>
<span><span class="fu">remotes</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"umccr/RNAsum#123"</span><span class="op">)</span> <span class="co"># PR 123</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="b-install-from-conda">B. Install from Conda<a class="anchor" aria-label="anchor" href="#b-install-from-conda"></a>
</h3>
<ul>
<li>
<strong>Conda</strong> package is available from the Anaconda <a href="https://anaconda.org/umccr/r-rnasum" class="external-link">umccr channel</a>:</li>
</ul>
<div class="sourceCode" id="cb6"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> create <span class="at">-n</span> rnasum <span class="at">-c</span> umccr <span class="at">-c</span> conda-forge <span class="at">-c</span> bioconda r-rnasum</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> activate rnasum</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="c-install-from-docker">C. Install from Docker<a class="anchor" aria-label="anchor" href="#c-install-from-docker"></a>
</h3>
<ul>
<li>
<strong>Docker</strong> image is available from the <a href="https://github.com/umccr/RNAsum/pkgs/container/rnasum" class="external-link">GitHub Container Registy</a>:</li>
</ul>
<div class="sourceCode" id="cb7"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="ex">docker</span> pull ghcr.io/umccr/rnasum:latest</span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="example-reports">Example reports<a class="anchor" aria-label="anchor" href="#example-reports"></a>
</h2>
<p><a href="https://doi.org/10.5281/zenodo.17353511" class="external-link"><img src="https://zenodo.org/badge/DOI/10.5281/zenodo.1234567.svg" alt="DOI"></a></p>
</div>
<div class="section level2">
<h2 id="workflow">Workflow<a class="anchor" aria-label="anchor" href="#workflow"></a>
</h2>
<p>The pipeline consists of five main components illustrated and briefly described below. For more details, see <a href="./inst/articles/workflow.html">workflow.md</a>.</p>
<p><img src="reference/figures/RNAsum_workflow_updated.png" width="100%"></p>
<ol style="list-style-type: decimal">
<li>Collect patient <strong>WTS data</strong> including per-gene read counts and gene fusions.</li>
<li>Add expression data from <strong><a href="#reference-data">reference cohorts</a></strong> to get an idea about the expression levels of genes of interest in other cancer patient cohorts. The read counts are normalised, transformed and converted into a scale that allows to present the patient’s expression measurements in the context of the reference cohorts.</li>
<li>Supply <strong>genome-based findings</strong> from whole-genome sequencing (WGS) data to focus on genes of interest and to provide additional evidence for dysregulation of mutated genes, or genes located within detected structural variants (SVs) or copy-number (CN) altered regions. <code>RNAsum</code> is designed to be compatible with WGS patient outputs generated from <code>umccrise</code>.</li>
<li>Collate results with knowledge derived from in-house resources and public databases to provide additional sources of evidence for clinical significance of altered genes e.g. to flag variants with clinical significance or potential druggable targets.</li>
<li>The final product is an interactive HTML report with searchable tables and plots presenting expression levels of the genes of interest. The report consists of several sections described <a href="./inst/articles/report_structure.html">here</a>.</li>
</ol>
</div>
<div class="section level2">
<h2 id="reference-data">Reference data<a class="anchor" aria-label="anchor" href="#reference-data"></a>
</h2>
<p>The reference expression data are available for <strong>33 cancer types</strong> and were derived from <a href="#external-reference-cohorts">external</a> (<a href="https://tcga-data.nci.nih.gov/" class="external-link">TCGA</a>) and <a href="#internal-reference-cohort">internal</a> (<a href="https://genomic-cancer-medicine.unimelb.edu.au/research/clinical-cancer-genomics" class="external-link">CCGCM</a>) resources.</p>
<div class="section level3">
<h3 id="external-reference-cohorts">External reference cohorts<a class="anchor" aria-label="anchor" href="#external-reference-cohorts"></a>
</h3>
<p>In order to explore expression changes in the patient, we have built a high-quality external cancer reference cohort.</p>
<p>Depending on the tissue from which the patient’s sample was taken, one of <strong>33 cancer datasets</strong> from TCGA can be used as a reference cohort for comparing expression changes in genes of interest of the patient. Additionally, 10 samples from each of the 33 TCGA datasets were combined to create the <strong><a href="./inst/articles/TCGA_projects_summary.md#pan-cancer-dataset">Pan-Cancer dataset</a></strong>, and for some cohorts <strong><a href="./inst/articles/TCGA_projects_summary.md#extended-datasets">extended sets</a></strong> are also available. All available datasets are listed in the <strong><a href="./inst/articles/TCGA_projects_summary.html">TCGA projects summary table</a></strong>. These datasets have been processed using methods described in the <a href="https://github.com/umccr/TCGA-data-harmonization/blob/master/expression/README.md#gdc-counts-data" class="external-link">TCGA-data-harmonization</a> repository. The dataset of interest can be specified by using one of the TCGA project IDs for the <code>RNAsum</code> <code>--dataset</code> argument (see <a href="#examples">Examples</a>).</p>
</div>
<div class="section level3">
<h3 id="internal-reference-cohort">Internal reference cohort<a class="anchor" aria-label="anchor" href="#internal-reference-cohort"></a>
</h3>
<p>The publicly available TCGA datasets are expected to demonstrate prominent <a href="https://www.ncbi.nlm.nih.gov/pubmed/20838408" class="external-link">batch effects</a> when compared to the in-house WTS data due to differences in applied experimental procedures and analytical pipelines. Moreover, TCGA data may include samples from tissue material of lower quality and cellularity compared to samples processed using local protocols. To address these issues, we have built a high-quality internal reference cohort processed using the same pipelines as input data (see <a href="./inst/articles/workflow.md#data-processing">data pre-processing</a>).</p>
<p>This internal reference set of <strong>40 pancreatic cancer samples</strong> is based on WTS data generated at <strong><a href="https://genomic-cancer-medicine.unimelb.edu.au/research/clinical-cancer-genomics" class="external-link">CCGCM</a></strong> and processed with the <strong>in-house RNA-seq</strong> pipeline to minimise potential batch effects between investigated samples and the reference cohort and to make sure the data are comparable. The internal reference cohort assembly is summarised in the <a href="https://github.com/umccr/Pancreatic-data-harmonization/tree/master/expression/in-house" class="external-link">Pancreatic-data-harmonization</a> repository.</p>
</div>
</div>
<div class="section level2">
<h2 id="usage">Usage<a class="anchor" aria-label="anchor" href="#usage"></a>
</h2>
<div class="sourceCode" id="cb8"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="va">rnasum_cli</span><span class="op">=</span><span class="va">$(</span><span class="ex">Rscript</span> <span class="at">-e</span> <span class="st">'cat(system.file("cli", package="RNAsum"))'</span><span class="va">)</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ln</span> <span class="at">-sf</span> <span class="st">"</span><span class="va">$rnasum_cli</span><span class="st">/rnasum.R"</span> <span class="st">"</span><span class="va">$rnasum_cli</span><span class="st">/rnasum"</span></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a><span class="bu">export</span> <span class="va">PATH</span><span class="op">=</span><span class="st">"</span><span class="va">$rnasum_cli</span><span class="st">:</span><span class="va">$PATH</span><span class="st">"</span></span></code></pre></div>
<div class="sourceCode" id="cb9"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span> rnasum <span class="sc">--</span>version</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="dv">2</span>.<span class="fl">0.0</span> </span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span> rnasum <span class="sc">--</span>help</span></code></pre></div>
<details><summary><strong>Options</strong>
</summary><pre><code>

--arriba_dir: Directory path to Arriba results containing fusions.pdf and fusions.tsv

--arriba_pdf: File path of Arriba PDF output

--arriba_tsv: File path of Arriba TSV output

--batch_rm: Remove batch-associated effects between datasets

--cn_gain: CN threshold value to classify genes within gained regions [def: 95]

--cn_loss: CN threshold value to classify genes within lost regions [def: 5]

--dataset: Dataset to be used as external reference cohort [def: PANCAN]

--dataset_name_incl: Include dataset in report name

--dragen_fusions: File path to DRAGEN RNA-seq 'fusion_candidates.final' output

--dragen_mapping_metrics: File path to DRAGEN RNA-seq 'mapping_metrics.csv' output

--dragen_wts_dir: Directory path to DRAGEN RNA-seq results

--drugs: Include drug matching section in report

--filter: Filter out low expressed genes

--immunogram: Include immunogram in report

--log: Log2 transform data before normalisation

--norm: Normalisation method

--pcgr_splice_vars: Include non-coding splice region variants reported in PCGR

--pcgr_tier: Tier threshold for reporting variants reported in PCGR [def: 4]

--pcgr_tiers_tsv: File path to PCGR 'snvs_indels.tiers.tsv' output

--project: Project name, used for annotation purposes only

--purple_gene_tsv: File path to PURPLE 'purple.cnv.gene.tsv' output

--report_dir: Directory path to output report

--salmon: File path to salmon 'quant.genes.sf' output

--kallisto: File path to kallisto 'abundance.tsv' output

--sample_name: Sample name to be presented in report

--sample_source: Type of investigated sample [def: -]

--save_tables: Save interactive summary tables as HTML

--scaling: Scaling for z-score transformation (gene-wise or group-wise) [def: gene-wise]

--subject_id: Subject ID

--sv_tsv: File path to text file with genes related to structural variation

--top_genes: Number of top ranked genes to be presented in report

--transform: Transformation method to be used when converting read counts [def: CPM]

--umccrise: Directory path of the corresponding WGS-related umccrise data

--version, -v: Print RNAsum version and exit

--help, -h: Show this help message and exit
</code></pre>
</details><p><strong>Note</strong></p>
<p>Human reference genome <strong><em><a href="https://www.ncbi.nlm.nih.gov/assembly/GCF_000001405.39" class="external-link">GRCh38</a></em></strong> (<em>Ensembl</em> based annotation version <strong><em>105</em></strong>) is used for gene annotation by default. GRCh37 is no longer supported.</p>
<div class="section level3">
<h3 id="examples">Examples<a class="anchor" aria-label="anchor" href="#examples"></a>
</h3>
<p>Below are <code>RNAsum</code> CLI commands for generating HTML reports under different data availability scenarios:</p>
<ol style="list-style-type: decimal">
<li><a href="#id_1-wts-and-wgs-data">WTS and WGS data</a></li>
<li><a href="#id_2-wts-data-only">WTS data only</a></li>
</ol>
<p><strong>Note</strong></p>
<ul>
<li>Example data is provided in the <code>/inst/rawdata/test_data</code> folder of the GitHub [repo][rnasum-gh].</li>
<li>The <code>RNAsum</code> runtime should be less than <strong>15 minutes</strong> using <strong>16GB RAM</strong> memory and <strong>1 CPU</strong>.</li>
</ul>
<div class="section level4">
<h4 id="id_1-wts-and-wgs-data">1. WTS and WGS data<a class="anchor" aria-label="anchor" href="#id_1-wts-and-wgs-data"></a>
</h4>
<p>This is the <strong>most frequent and preferred case</strong>, in which the <a href="#wgs">WGS</a>-based findings will be used as a primary source for expression profile prioritisation. The genome-based results can be incorporated into the report by specifying the location of the corresponding output files (e.g. results from <code>PCGR</code>, <code>PURPLE</code>, and <code>Manta</code>). The <strong><code>Mutated genes</code></strong>, <strong><code>Structural variants</code></strong> and <strong><code>CN altered genes</code></strong> report sections will contain information about expression levels of the mutated genes, genes located within detected SVs and CN altered regions, respectively. The results in the <strong><code>Fusion genes</code></strong> section will be ordered based on the evidence from genome-based data. A subset of the TCGA pancreatic adenocarcinoma dataset is used as reference cohort (<code>--dataset TEST</code>).</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> <span class="va">$rnasum_cli</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a><span class="ex">rnasum</span> <span class="dt">\</span></span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">--sample_name</span> test_sample_WTS <span class="dt">\</span></span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">--dataset</span> TEST <span class="dt">\</span></span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">--salmon</span> <span class="st">"</span><span class="va">$PWD</span><span class="st">/../rawdata/test_data/dragen/TEST.quant.genes.sf"</span> <span class="dt">\</span></span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">--arriba_pdf</span> <span class="st">"</span><span class="va">$PWD</span><span class="st">/../rawdata/test_data/dragen/arriba/fusions.pdf"</span> <span class="dt">\</span></span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">--arriba_tsv</span> <span class="st">"</span><span class="va">$PWD</span><span class="st">/../rawdata/test_data/dragen/arriba/fusions.tsv"</span>  <span class="dt">\</span></span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">--dragen_fusions</span> <span class="st">"</span><span class="va">$PWD</span><span class="st">/../rawdata/test_data/dragen/test_sample_WTS.fusion_candidates.final"</span>  <span class="dt">\</span></span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">--pcgr_tiers_tsv</span> <span class="st">"</span><span class="va">$PWD</span><span class="st">/../rawdata/test_data/umccrised/test_sample_WGS/small_variants/TEST-somatic.pcgr.snvs_indels.tiers.tsv"</span> <span class="dt">\</span></span>
<span id="cb10-11"><a href="#cb10-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">--purple_gene_tsv</span> <span class="st">"</span><span class="va">$PWD</span><span class="st">/../rawdata/test_data/umccrised/test_sample_WGS/purple/TEST.purple.cnv.gene.tsv"</span> <span class="dt">\</span></span>
<span id="cb10-12"><a href="#cb10-12" aria-hidden="true" tabindex="-1"></a>  <span class="at">--sv_tsv</span> <span class="st">"</span><span class="va">$PWD</span><span class="st">/../rawdata/test_data/umccrised/test_sample_WGS/structural/TEST-manta.tsv"</span> <span class="dt">\</span></span>
<span id="cb10-13"><a href="#cb10-13" aria-hidden="true" tabindex="-1"></a>  <span class="at">--report_dir</span> <span class="st">"</span><span class="va">$PWD</span><span class="st">/../rawdata/test_data/RNAsum"</span> <span class="dt">\</span></span>
<span id="cb10-14"><a href="#cb10-14" aria-hidden="true" tabindex="-1"></a>  <span class="at">--save_tables</span> FALSE <span class="dt">\</span></span>
<span id="cb10-15"><a href="#cb10-15" aria-hidden="true" tabindex="-1"></a>  <span class="at">--filter</span> TRUE</span></code></pre></div>
<p>The HTML report <code>test_sample_WTS.RNAsum.html</code> will be created in the <code>inst/rawdata/test_data/dragen/RNAsum</code> folder.</p>
</div>
<div class="section level4">
<h4 id="id_2-wts-data-only">2. WTS data only<a class="anchor" aria-label="anchor" href="#id_2-wts-data-only"></a>
</h4>
<p>In this scenario, only <a href="#wts">WTS</a> data will be used and only expression levels of key <strong><a href="./inst/articles/workflow.md#key-cancer-genes"><code>Cancer genes</code></a></strong>, <strong><code>Fusion genes</code></strong>, <strong><code>Immune markers</code></strong> and homologous recombination deficiency genes (<strong><code>HRD genes</code></strong>) will be reported. Moreover, gene fusions reported in the <code>Fusion genes</code> report section will not contain information about evidence from genome-based data. A subset of the TCGA pancreatic adenocarcinoma dataset is used as the reference cohort (<code>--dataset TEST</code>).</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> <span class="va">$rnasum_cli</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a><span class="ex">rnasum</span> <span class="dt">\</span></span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">--sample_name</span> test_sample_WTS <span class="dt">\</span></span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">--dataset</span> TEST <span class="dt">\</span></span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">--salmon</span> <span class="st">"</span><span class="va">$PWD</span><span class="st">/../rawdata/test_data/dragen/TEST.quant.genes.sf"</span> <span class="dt">\</span></span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">--arriba_pdf</span> <span class="st">"</span><span class="va">$PWD</span><span class="st">/../rawdata/test_data/dragen/arriba/fusions.pdf"</span> <span class="dt">\</span></span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">--arriba_tsv</span> <span class="st">"</span><span class="va">$PWD</span><span class="st">/../rawdata/test_data/dragen/arriba/fusions.tsv"</span>  <span class="dt">\</span></span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">--report_dir</span> <span class="st">"</span><span class="va">$PWD</span><span class="st">/../rawdata/test_data/RNAsum"</span> <span class="dt">\</span></span>
<span id="cb11-10"><a href="#cb11-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">--save_tables</span> FALSE <span class="dt">\</span></span>
<span id="cb11-11"><a href="#cb11-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">--filter</span> TRUE</span></code></pre></div>
<p>The output HTML report <code>test_sample_WTS.RNAsum.html</code> will be created in the <code>inst/rawdata/test_data/dragen/RNAsum</code> folder.</p>
</div>
</div>
<div class="section level3">
<h3 id="output">Output<a class="anchor" aria-label="anchor" href="#output"></a>
</h3>
<p>The pipeline generates a HTML <strong><em>Patient Transcriptome Summary</em></strong> <strong><a href="#report">report</a></strong> and a <a href="#results">results</a> folder:</p>
<pre class="text"><code>|
|____&lt;output&gt;
  |____&lt;SampleName&gt;.&lt;output&gt;.html
  |____results
    |____exprTables
    |____glanceExprPlots
    |____...</code></pre>
<div class="section level4">
<h4 id="report">Report<a class="anchor" aria-label="anchor" href="#report"></a>
</h4>
<p>The generated HTML report includes searchable tables and interactive plots presenting expression levels of altered genes, as well as links to public resources describing the genes of interest.</p>
<p>Detailed description of the report structure, including result prioritisation and visualisation is available <a href="./inst/articles/report_structure.html">here</a>.</p>
</div>
<div class="section level4">
<h4 id="results">Results<a class="anchor" aria-label="anchor" href="#results"></a>
</h4>
<p>The <code>results</code> folder contains intermediate files, including plots and tables that are presented in the HTML report.</p>
</div>
<div class="section level4">
<h4 id="code-of-conduct">Code of Conduct<a class="anchor" aria-label="anchor" href="#code-of-conduct"></a>
</h4>
<p>The code of conduct can be accessed <a href="./CODE_OF_CONDUCT.html">here</a>.</p>
</div>
<div class="section level4">
<h4 id="citation">Citation<a class="anchor" aria-label="anchor" href="#citation"></a>
</h4>
<p>To cite package ‘RNAsum’ in publications use:</p>
<blockquote>
<p>Kanwal S, Marzec J, Diakumis P, Hofmann O, Grimmond S (2024). “RNAsum: An R package to comprehensively post-process, summarise and visualise genomics and transcriptomics data.” version 1.1.0, <a href="https://umccr.github.io/RNAsum/" class="external-link uri">https://umccr.github.io/RNAsum/</a>.</p>
</blockquote>
<p>A BibTeX entry for LaTeX users is</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>  <span class="sc">@</span>Unpublished{,</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>    title <span class="ot">=</span> {RNAsum<span class="sc">:</span> An R package to comprehensively post<span class="sc">-</span>process, summarise and visualise genomics and transcriptomics data},</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>    author <span class="ot">=</span> {Sehrish Kanwal and Jacek Marzec and Peter Diakumis and Oliver Hofmann and Sean Grimmond},</span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a>    year <span class="ot">=</span> {<span class="dv">2024</span>},</span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a>    note <span class="ot">=</span> {version <span class="dv">1</span>.<span class="fl">1.0</span>},</span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a>    url <span class="ot">=</span> {https<span class="sc">:</span><span class="er">//</span>umccr.github.io<span class="sc">/</span>RNAsum<span class="sc">/</span>},</span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a>  }</span></code></pre></div>
</div>
</div>
</div>
</div>

  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/umccr/RNAsum/" class="external-link">Browse source code</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>

<div class="community">
<h2 data-toc-skip>Community</h2>
<ul class="list-unstyled">
<li><a href="CONTRIBUTING.html">Contributing guide</a></li>
<li><a href="CODE_OF_CONDUCT.html">Code of conduct</a></li>
</ul>
</div>

<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing RNAsum</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Sehrish Kanwal <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0002-5044-4692" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Jacek Marzec <br><small class="roles"> Author </small>  </li>
</ul>
</div>



  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Sehrish Kanwal, Jacek Marzec.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
