<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="RNA-seq reporting workflow designed to post-process,
    summarise and visualise an output from bcbio-nextgen RNA-seq or Dragen
    RNA pipelines. Its main application is to complement genome-based
    findings and provide additional evidence for detected alterations.">
<title>A Robust Tool for Performing Transcriptome Profiling on Cancer Data • RNAsum</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="A Robust Tool for Performing Transcriptome Profiling on Cancer Data">
<meta property="og:description" content="RNA-seq reporting workflow designed to post-process,
    summarise and visualise an output from bcbio-nextgen RNA-seq or Dragen
    RNA pipelines. Its main application is to complement genome-based
    findings and provide additional evidence for detected alterations.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="index.html">RNAsum</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.5.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="articles/development.html">Development</a>
    <a class="dropdown-item" href="articles/report_structure.html">RNAsum Sections</a>
    <a class="dropdown-item" href="articles/tcga_projects_summary.html">TCGA Projects</a>
    <a class="dropdown-item" href="articles/workflow.html">Workflow Details</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/umccr/RNAsum/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><ul>
<li>
<a href="#rnasum">RNAsum</a>
<ul>
<li><a href="#installation">Installation</a></li>
<li><a href="#workflow">Workflow</a></li>
<li>
<a href="#reference-data">Reference data</a>
<ul>
<li><a href="#external-reference-cohorts">External reference cohorts</a></li>
<li><a href="#internal-reference-cohort">Internal reference cohort</a></li>
</ul>
</li>
<li>
<a href="#input-data">Input data</a>
<ul>
<li><a href="#wts">WTS</a></li>
<li><a href="#wgs">WGS</a></li>
</ul>
</li>
<li>
<a href="#usage">Usage</a>
<ul>
<li><a href="#examples">Examples</a></li>
<li><a href="#output">Output</a></li>
</ul>
</li>
</ul>
</li>
</ul>
<!-- README.md is generated from README.Rmd. Please edit that file --><div class="section level1">
<div class="page-header"><h1 id="rnasum">RNAsum<a class="anchor" aria-label="anchor" href="#rnasum"></a>
</h1></div>
<p><code>RNAsum</code> is an R package that can post-process, summarise and visualise outputs primarily from <a href="https://sapac.illumina.com/products/by-type/informatics-products/basespace-sequence-hub/apps/edico-genome-inc-dragen-rna-pipeline.html" class="external-link">DRAGEN RNA</a> pipelines. Its main application is to complement genome-based findings from the <a href="https://github.com/umccr/umccrise" class="external-link">umccrise</a> pipeline and to provide additional evidence for detected alterations.</p>
<p><strong>DOCS</strong>: <a href="https://umccr.github.io/RNAsum" class="external-link uri">https://umccr.github.io/RNAsum</a></p>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<ul>
<li>
<strong>R</strong> package can be installed directly from the <a href="https://github.com/umccr/RNAsum" class="external-link">GitHub source</a>:</li>
</ul>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"umccr/RNAsum"</span><span class="op">)</span> <span class="co"># latest master commit</span></span>
<span><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"umccr/RNAsum@v0.0.X"</span><span class="op">)</span> <span class="co"># version 0.0.X</span></span>
<span><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"umccr/RNAsum@abcde"</span><span class="op">)</span> <span class="co"># commit abcde</span></span>
<span><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"umccr/RNAsum#123"</span><span class="op">)</span> <span class="co"># PR 123</span></span></code></pre></div>
<ul>
<li>
<strong>Conda</strong> package is available from the Anaconda <a href="https://anaconda.org/umccr/r-rnasum" class="external-link">umccr channel</a>:</li>
</ul>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> install r-rnasum==0.0.X <span class="at">-c</span> umccr <span class="at">-c</span> conda-forge <span class="at">-c</span> bioconda</span></code></pre></div>
<ul>
<li>
<strong>Docker</strong> image is available from the <a href="https://github.com/umccr/RNAsum/pkgs/container/rnasum" class="external-link">GitHub Container Registy</a>:</li>
</ul>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="ex">docker</span> pull ghcr.io/umccr/rnasum:latest</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="workflow">Workflow<a class="anchor" aria-label="anchor" href="#workflow"></a>
</h2>
<p>The pipeline consists of five main components illustrated and briefly described below. For more details, see <a href="/workflow.html">workflow.md</a>.</p>
<p><img src="reference/figures/RNAsum_workflow.png" width="100%"></p>
<ol style="list-style-type: decimal">
<li>Collect patient <strong>WTS data</strong> from the <a href="https://sapac.illumina.com/products/by-type/informatics-products/basespace-sequence-hub/apps/edico-genome-inc-dragen-rna-pipeline.html" class="external-link">DRAGEN RNA</a> pipeline including per-gene read counts and gene fusions.</li>
<li>Add expression data from <strong><a href="#reference-data">reference cohorts</a></strong> to get an idea about the expression levels of genes of interest in other cancer patient cohorts. The read counts are normalised, transformed and converted into a scale that allows to present the patient’s expression measurements in the context of the reference cohorts.</li>
<li>Supply <strong>genome-based findings</strong> from whole-genome sequencing (WGS) data to focus on genes of interest and to provide additional evidence for dysregulation of mutated genes, or genes located within detected structural variants (SVs) or copy-number (CN) altered regions. <code>RNAsum</code> is designed to be compatible with WGS patient outputs generated from <code>umccrise</code>.</li>
<li>Collate results with knowledge derived from in-house resources and public databases to provide additional sources of evidence for clinical significance of altered genes e.g. to flag variants with clinical significance or potential druggable targets.</li>
<li>The final product is an interactive HTML report with searchable tables and plots presenting expression levels of the genes of interest. The report consists of several sections described <a href="./articles/report_structure.html">here</a>.</li>
</ol>
</div>
<div class="section level2">
<h2 id="reference-data">Reference data<a class="anchor" aria-label="anchor" href="#reference-data"></a>
</h2>
<p>The reference expression data are available for <strong>33 cancer types</strong> and were derived from <a href="#external-reference-cohorts">external</a> (<a href="https://tcga-data.nci.nih.gov/" class="external-link">TCGA</a>) and <a href="#internal-reference-cohort">internal</a> (<a href="https://research.unimelb.edu.au/centre-for-cancer-research/our-research/precision-oncology-research-group" class="external-link">UMCCR</a>) resources.</p>
<div class="section level3">
<h3 id="external-reference-cohorts">External reference cohorts<a class="anchor" aria-label="anchor" href="#external-reference-cohorts"></a>
</h3>
<p>In order to explore expression changes in the patient, we have built a high-quality pancreatic cancer reference cohort.</p>
<p>Depending on the tissue from which the patient’s sample was taken, one of <strong>33 cancer datasets</strong> from TCGA can be used as a reference cohort for comparing expression changes in genes of interest of the patient. Additionally, 10 samples from each of the 33 TCGA datasets were combined to create the <strong><a href="./articles/tcga_projects_summary.md#pan-cancer-dataset">Pan-Cancer dataset</a></strong>, and for some cohorts <strong><a href="./articles/tcga_projects_summary.md#extended-datasets">extended sets</a></strong> are also available. All available datasets are listed in the <strong><a href="./articles/tcga_projects_summary.html">TCGA projects summary table</a></strong>. These datasets have been processed using methods described in the <a href="https://github.com/umccr/TCGA-data-harmonization/blob/master/expression/README.md#gdc-counts-data" class="external-link">TCGA-data-harmonization</a> repository. The dataset of interest can be specified by using one of the TCGA project IDs for the <code>RNAsum</code> <code>--dataset</code> argument (see <a href="#examples">Examples</a>).</p>
</div>
<div class="section level3">
<h3 id="internal-reference-cohort">Internal reference cohort<a class="anchor" aria-label="anchor" href="#internal-reference-cohort"></a>
</h3>
<p>The publicly available TCGA datasets are expected to demonstrate prominent <a href="https://www.ncbi.nlm.nih.gov/pubmed/20838408" class="external-link">batch effects</a> when compared to the in-house WTS data due to differences in applied experimental procedures and analytical pipelines. Moreover, TCGA data may include samples from tissue material of lower quality and cellularity compared to samples processed using local protocols. To address these issues, we have built a high-quality internal reference cohort processed using the same pipelines as input data (see <a href="./articles/workflow.md#data-processing">data pre-processing</a>).</p>
<p>This internal reference set of <strong>40 pancreatic cancer samples</strong> is based on WTS data generated at <strong><a href="https://research.unimelb.edu.au/centre-for-cancer-research/our-research/precision-oncology-research-group" class="external-link">UMCCR</a></strong> and processed with the <strong>bcbio-nextgen RNA-seq</strong> pipeline to minimise potential batch effects between investigated samples and the reference cohort and to make sure the data are comparable. The internal reference cohort assembly is summarised in the <a href="https://github.com/umccr/Pancreatic-data-harmonization/tree/master/expression/in-house" class="external-link">Pancreatic-data-harmonization</a> repository.</p>
<p><strong>Note</strong></p>
<p>There are two rationales for using the internal reference cohort:</p>
<ol style="list-style-type: decimal">
<li>In case of <strong>pancreatic cancer samples</strong> this cohort is used:
<ul>
<li>in <strong><em>batch effects correction</em></strong>
</li>
<li>as a reference point for <strong><em>comparing per-gene expression levels</em></strong> observed in the data of the patient of interest and data from other pancreatic cancer patients.</li>
</ul>
</li>
<li>In case of samples from <strong>any cancer type</strong> the data from the internal reference cohort is used in the <strong><em>batch effects correction</em></strong> procedure performed to minimise technical-related variation in the data.</li>
</ol>
</div>
</div>
<div class="section level2">
<h2 id="input-data">Input data<a class="anchor" aria-label="anchor" href="#input-data"></a>
</h2>
<p><code>RNAsum</code> accepts <a href="#wts">WTS</a> data processed by the <code>DRAGEN RNA</code> pipeline. Additionally, the WTS data can be integrated with <a href="#wgs">WGS</a>-based data processed using the <code>umccrise</code> pipeline. In the latter case, the genome-based findings from the corresponding patient sample are incorporated into the report and are used as a primary source for expression profile prioritisation.</p>
<div class="section level3">
<h3 id="wts">WTS<a class="anchor" aria-label="anchor" href="#wts"></a>
</h3>
<p>The only required WTS input data are <strong>read counts</strong> provided in a quantification file from the <code>DRAGEN RNA</code> pipeline.</p>
<div class="section level4">
<h4 id="dragen-rna">DRAGEN RNA<a class="anchor" aria-label="anchor" href="#dragen-rna"></a>
</h4>
<p>The table below lists all input data accepted in <code>RNAsum</code>:</p>
<table class="table">
<colgroup>
<col width="12%">
<col width="49%">
<col width="34%">
<col width="3%">
</colgroup>
<thead><tr class="header">
<th>Input file</th>
<th>Tool</th>
<th>Example</th>
<th>Required</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Quantified transcript <strong>abundances</strong>
</td>
<td>
<a href="https://salmon.readthedocs.io/en/latest/salmon.html" class="external-link">salmon</a> (<a href="https://salmon.readthedocs.io/en/latest/file_formats.html#fileformats" class="external-link">description</a>)</td>
<td><a href="/inst/rawdata/test_data/dragen/TEST.quant.sf">TEST.quant.sf</a></td>
<td><strong>Yes</strong></td>
</tr>
<tr class="even">
<td>
<strong>Fusion gene</strong> list</td>
<td><a href="https://sapac.illumina.com/products/by-type/informatics-products/basespace-sequence-hub/apps/edico-genome-inc-dragen-rna-pipeline.html" class="external-link">DRAGEN RNA</a></td>
<td><a href="/inst/rawdata/test_data/dragen/test_sample_WTS.fusion_candidates.final">TEST.fusion_candidates.final</a></td>
<td>No</td>
</tr>
</tbody>
</table>
<p>These files are expected to be organised in the following structure:</p>
<pre class="text"><code>|
|____&lt;SampleName&gt;
  |____&lt;SampleName&gt;quant.sf
  |____&lt;SampleName&gt;.fusion_candidates.final</code></pre>
</div>
</div>
<div class="section level3">
<h3 id="wgs">WGS<a class="anchor" aria-label="anchor" href="#wgs"></a>
</h3>
<p><code>RNAsum</code> is designed to be compatible with WGS outputs generated from <code>umccrise</code>.</p>
<p>The table below lists all input data accepted in <code>RNAsum</code>:</p>
<table class="table">
<colgroup>
<col width="7%">
<col width="32%">
<col width="55%">
<col width="4%">
</colgroup>
<thead><tr class="header">
<th>Input file</th>
<th>Tool</th>
<th>Example</th>
<th>Required</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><strong>SNVs/Indels</strong></td>
<td><a href="https://github.com/sigven/pcgr" class="external-link">PCGR</a></td>
<td><a href="/inst/rawdata/test_data/umccrised/test_sample_WGS/small_variants/pcgr.snvs_indels.tiers.tsv">pcgr.snvs_indels.tiers.tsv</a></td>
<td>No</td>
</tr>
<tr class="even">
<td><strong>CNVs</strong></td>
<td><a href="https://github.com/hartwigmedical/hmftools/tree/master/purple" class="external-link">PURPLE</a></td>
<td><a href="/inst/rawdata/test_data/umccrised/test_sample_WGS/purple/purple.gene.cnv">purple.cnv.gene.tsv</a></td>
<td>No</td>
</tr>
<tr class="odd">
<td><strong>SVs</strong></td>
<td><a href="https://github.com/Illumina/manta" class="external-link">Manta</a></td>
<td><a href="/inst/rawdata/test_data/umccrised/test_sample_WGS/structural/sv-prioritize-manta.tsv">sv-prioritize-manta.tsv</a></td>
<td>No</td>
</tr>
</tbody>
</table>
<p>These files are expected to be organised in the following structure:</p>
<pre class="text"><code>|
|____umccrised
  |____&lt;SampleName&gt;
    |____pcgr
    | |____&lt;SampleName&gt;-somatic.pcgr.snvs_indels.tiers.tsv
    |____purple
    | |____&lt;SampleName&gt;.purple.gene.cnv
    |____structural
      |____&lt;SampleName&gt;-manta.tsv</code></pre>
</div>
</div>
<div class="section level2">
<h2 id="usage">Usage<a class="anchor" aria-label="anchor" href="#usage"></a>
</h2>
<div class="sourceCode" id="cb6"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="va">rnasum_cli</span><span class="op">=</span><span class="va">$(</span><span class="ex">Rscript</span> <span class="at">-e</span> <span class="st">'x = system.file("cli", package = "RNAsum"); cat(x, "\n")'</span> <span class="kw">|</span> <span class="fu">xargs</span><span class="va">)</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="bu">export</span> <span class="va">PATH</span><span class="op">=</span><span class="st">"</span><span class="va">${rnasum_cli}</span><span class="st">:</span><span class="va">${PATH}</span><span class="st">"</span></span></code></pre></div>
<div class="sourceCode" id="cb7"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span> rnasum.R <span class="sc">--</span>version</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>rnasum.R x.x.x</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span> rnasum.R <span class="sc">--</span>help</span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a>Usage</span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a><span class="sc">==</span><span class="er">===</span></span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a> </span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a><span class="er">/</span>Library<span class="sc">/</span>Frameworks<span class="sc">/</span>R.framework<span class="sc">/</span>Versions<span class="sc">/</span><span class="fl">4.3</span><span class="sc">-</span>arm64<span class="sc">/</span>Resources<span class="sc">/</span>library<span class="sc">/</span>RNAsum<span class="sc">/</span>cli<span class="sc">/</span>rnasum.R [options]</span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-10"><a href="#cb7-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-11"><a href="#cb7-11" aria-hidden="true" tabindex="-1"></a>Options</span>
<span id="cb7-12"><a href="#cb7-12" aria-hidden="true" tabindex="-1"></a><span class="sc">==</span><span class="er">=====</span></span>
<span id="cb7-13"><a href="#cb7-13" aria-hidden="true" tabindex="-1"></a><span class="sc">--</span>arriba_pdf<span class="ot">=</span>ARRIBA_PDF</span>
<span id="cb7-14"><a href="#cb7-14" aria-hidden="true" tabindex="-1"></a>        File path of Arriba PDF output</span>
<span id="cb7-15"><a href="#cb7-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-16"><a href="#cb7-16" aria-hidden="true" tabindex="-1"></a><span class="sc">--</span>arriba_tsv<span class="ot">=</span>ARRIBA_TSV</span>
<span id="cb7-17"><a href="#cb7-17" aria-hidden="true" tabindex="-1"></a>        File path of Arriba TSV output</span>
<span id="cb7-18"><a href="#cb7-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-19"><a href="#cb7-19" aria-hidden="true" tabindex="-1"></a><span class="sc">--</span>batch_rm<span class="ot">=</span>BATCH_RM</span>
<span id="cb7-20"><a href="#cb7-20" aria-hidden="true" tabindex="-1"></a>        Remove batch<span class="sc">-</span>associated effects between datasets? [def<span class="sc">:</span> <span class="cn">TRUE</span>]</span>
<span id="cb7-21"><a href="#cb7-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-22"><a href="#cb7-22" aria-hidden="true" tabindex="-1"></a><span class="sc">--</span>cn_gain<span class="ot">=</span>CN_GAIN</span>
<span id="cb7-23"><a href="#cb7-23" aria-hidden="true" tabindex="-1"></a>        CN threshold value to classify genes within gained regions [def<span class="sc">:</span> <span class="dv">95</span>]</span>
<span id="cb7-24"><a href="#cb7-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-25"><a href="#cb7-25" aria-hidden="true" tabindex="-1"></a><span class="sc">--</span>cn_loss<span class="ot">=</span>CN_LOSS</span>
<span id="cb7-26"><a href="#cb7-26" aria-hidden="true" tabindex="-1"></a>        CN threshold value to classify genes within lost regions [def<span class="sc">:</span> <span class="dv">5</span>]</span>
<span id="cb7-27"><a href="#cb7-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-28"><a href="#cb7-28" aria-hidden="true" tabindex="-1"></a><span class="sc">--</span>dataset<span class="ot">=</span>DATASET</span>
<span id="cb7-29"><a href="#cb7-29" aria-hidden="true" tabindex="-1"></a>        Dataset to be used as external reference cohort [def<span class="sc">:</span> PANCAN]</span>
<span id="cb7-30"><a href="#cb7-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-31"><a href="#cb7-31" aria-hidden="true" tabindex="-1"></a><span class="sc">--</span>dataset_name_incl<span class="ot">=</span>DATASET_NAME_INCL</span>
<span id="cb7-32"><a href="#cb7-32" aria-hidden="true" tabindex="-1"></a>        Include dataset <span class="cf">in</span> report name? [def<span class="sc">:</span> <span class="cn">FALSE</span>]</span>
<span id="cb7-33"><a href="#cb7-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-34"><a href="#cb7-34" aria-hidden="true" tabindex="-1"></a><span class="sc">--</span>dragen_fusions<span class="ot">=</span>DRAGEN_FUSIONS</span>
<span id="cb7-35"><a href="#cb7-35" aria-hidden="true" tabindex="-1"></a>        File path to DRAGEN RNA<span class="sc">-</span>seq <span class="st">'fusion_candidates.final'</span> output</span>
<span id="cb7-36"><a href="#cb7-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-37"><a href="#cb7-37" aria-hidden="true" tabindex="-1"></a><span class="sc">--</span>drugs<span class="ot">=</span>DRUGS</span>
<span id="cb7-38"><a href="#cb7-38" aria-hidden="true" tabindex="-1"></a>        Include drug matching section <span class="cf">in</span> report? [def<span class="sc">:</span> <span class="cn">FALSE</span>]</span>
<span id="cb7-39"><a href="#cb7-39" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-40"><a href="#cb7-40" aria-hidden="true" tabindex="-1"></a><span class="sc">--</span>filter<span class="ot">=</span>FILTER</span>
<span id="cb7-41"><a href="#cb7-41" aria-hidden="true" tabindex="-1"></a>        Filter out low expressed genes? [def<span class="sc">:</span> <span class="cn">TRUE</span>]</span>
<span id="cb7-42"><a href="#cb7-42" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-43"><a href="#cb7-43" aria-hidden="true" tabindex="-1"></a><span class="sc">--</span>hide_code_btn<span class="ot">=</span>HIDE_CODE_BTN</span>
<span id="cb7-44"><a href="#cb7-44" aria-hidden="true" tabindex="-1"></a>        Hide <span class="st">"Code"</span> button above code chunks <span class="cf">in</span> report? [def<span class="sc">:</span> <span class="cn">TRUE</span>]</span>
<span id="cb7-45"><a href="#cb7-45" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-46"><a href="#cb7-46" aria-hidden="true" tabindex="-1"></a><span class="sc">--</span>immunogram<span class="ot">=</span>IMMUNOGRAM</span>
<span id="cb7-47"><a href="#cb7-47" aria-hidden="true" tabindex="-1"></a>        Include immunogram <span class="cf">in</span> report? [def<span class="sc">:</span> <span class="cn">FALSE</span>]</span>
<span id="cb7-48"><a href="#cb7-48" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-49"><a href="#cb7-49" aria-hidden="true" tabindex="-1"></a><span class="sc">--</span>log<span class="ot">=</span>LOG</span>
<span id="cb7-50"><a href="#cb7-50" aria-hidden="true" tabindex="-1"></a>        Log2 transform data before normalisation? [def<span class="sc">:</span> <span class="cn">TRUE</span>]</span>
<span id="cb7-51"><a href="#cb7-51" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-52"><a href="#cb7-52" aria-hidden="true" tabindex="-1"></a><span class="sc">--</span>manta_tsv<span class="ot">=</span>MANTA_TSV</span>
<span id="cb7-53"><a href="#cb7-53" aria-hidden="true" tabindex="-1"></a>        File path to umccrise <span class="st">'manta.tsv'</span> output</span>
<span id="cb7-54"><a href="#cb7-54" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-55"><a href="#cb7-55" aria-hidden="true" tabindex="-1"></a><span class="sc">--</span>norm<span class="ot">=</span>NORM</span>
<span id="cb7-56"><a href="#cb7-56" aria-hidden="true" tabindex="-1"></a>        Normalisation method</span>
<span id="cb7-57"><a href="#cb7-57" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-58"><a href="#cb7-58" aria-hidden="true" tabindex="-1"></a><span class="sc">--</span>pcgr_splice_vars<span class="ot">=</span>PCGR_SPLICE_VARS</span>
<span id="cb7-59"><a href="#cb7-59" aria-hidden="true" tabindex="-1"></a>        Include non<span class="sc">-</span>coding splice region variants reported <span class="cf">in</span> PCGR? [def<span class="sc">:</span> <span class="cn">TRUE</span>]</span>
<span id="cb7-60"><a href="#cb7-60" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-61"><a href="#cb7-61" aria-hidden="true" tabindex="-1"></a><span class="sc">--</span>pcgr_tier<span class="ot">=</span>PCGR_TIER</span>
<span id="cb7-62"><a href="#cb7-62" aria-hidden="true" tabindex="-1"></a>        Tier threshold <span class="cf">for</span> reporting variants reported <span class="cf">in</span> PCGR [def<span class="sc">:</span> <span class="dv">4</span>]</span>
<span id="cb7-63"><a href="#cb7-63" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-64"><a href="#cb7-64" aria-hidden="true" tabindex="-1"></a><span class="sc">--</span>pcgr_tiers_tsv<span class="ot">=</span>PCGR_TIERS_TSV</span>
<span id="cb7-65"><a href="#cb7-65" aria-hidden="true" tabindex="-1"></a>        File path to PCGR <span class="st">'snvs_indels.tiers.tsv'</span> output</span>
<span id="cb7-66"><a href="#cb7-66" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-67"><a href="#cb7-67" aria-hidden="true" tabindex="-1"></a><span class="sc">--</span>project<span class="ot">=</span>PROJECT</span>
<span id="cb7-68"><a href="#cb7-68" aria-hidden="true" tabindex="-1"></a>        Project name</span>
<span id="cb7-69"><a href="#cb7-69" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-70"><a href="#cb7-70" aria-hidden="true" tabindex="-1"></a><span class="sc">--</span>purple_gene_tsv<span class="ot">=</span>PURPLE_GENE_TSV</span>
<span id="cb7-71"><a href="#cb7-71" aria-hidden="true" tabindex="-1"></a>        File path to PURPLE <span class="st">'purple.cnv.gene.tsv'</span> output</span>
<span id="cb7-72"><a href="#cb7-72" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-73"><a href="#cb7-73" aria-hidden="true" tabindex="-1"></a><span class="sc">--</span>report_dir<span class="ot">=</span>REPORT_DIR</span>
<span id="cb7-74"><a href="#cb7-74" aria-hidden="true" tabindex="-1"></a>        Directory path to output report</span>
<span id="cb7-75"><a href="#cb7-75" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-76"><a href="#cb7-76" aria-hidden="true" tabindex="-1"></a><span class="sc">--</span>salmon<span class="ot">=</span>SALMON</span>
<span id="cb7-77"><a href="#cb7-77" aria-hidden="true" tabindex="-1"></a>        File path to salmon <span class="st">'quant.sf'</span> output</span>
<span id="cb7-78"><a href="#cb7-78" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-79"><a href="#cb7-79" aria-hidden="true" tabindex="-1"></a><span class="sc">--</span>sample_name<span class="ot">=</span>SAMPLE_NAME</span>
<span id="cb7-80"><a href="#cb7-80" aria-hidden="true" tabindex="-1"></a>        Sample name to be presented <span class="cf">in</span> report</span>
<span id="cb7-81"><a href="#cb7-81" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-82"><a href="#cb7-82" aria-hidden="true" tabindex="-1"></a><span class="sc">--</span>sample_source<span class="ot">=</span>SAMPLE_SOURCE</span>
<span id="cb7-83"><a href="#cb7-83" aria-hidden="true" tabindex="-1"></a>        Type of investigated sample [def<span class="sc">:</span> <span class="sc">-</span>]</span>
<span id="cb7-84"><a href="#cb7-84" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-85"><a href="#cb7-85" aria-hidden="true" tabindex="-1"></a><span class="sc">--</span>save_tables<span class="ot">=</span>SAVE_TABLES</span>
<span id="cb7-86"><a href="#cb7-86" aria-hidden="true" tabindex="-1"></a>        Save interactive summary tables as HTML? [def<span class="sc">:</span> <span class="cn">TRUE</span>]</span>
<span id="cb7-87"><a href="#cb7-87" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-88"><a href="#cb7-88" aria-hidden="true" tabindex="-1"></a><span class="sc">--</span>scaling<span class="ot">=</span>SCALING</span>
<span id="cb7-89"><a href="#cb7-89" aria-hidden="true" tabindex="-1"></a>        Scaling <span class="cf">for</span> z<span class="sc">-</span>score <span class="fu">transformation</span> (gene<span class="sc">-</span>wise or group<span class="sc">-</span>wise) [def<span class="sc">:</span> gene<span class="sc">-</span>wise]</span>
<span id="cb7-90"><a href="#cb7-90" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-91"><a href="#cb7-91" aria-hidden="true" tabindex="-1"></a><span class="sc">--</span>subject_id<span class="ot">=</span>SUBJECT_ID</span>
<span id="cb7-92"><a href="#cb7-92" aria-hidden="true" tabindex="-1"></a>        Subject ID</span>
<span id="cb7-93"><a href="#cb7-93" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-94"><a href="#cb7-94" aria-hidden="true" tabindex="-1"></a><span class="sc">--</span>top_genes<span class="ot">=</span>TOP_GENES</span>
<span id="cb7-95"><a href="#cb7-95" aria-hidden="true" tabindex="-1"></a>        Number of top ranked genes to be presented <span class="cf">in</span> report</span>
<span id="cb7-96"><a href="#cb7-96" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-97"><a href="#cb7-97" aria-hidden="true" tabindex="-1"></a><span class="sc">--</span>transform<span class="ot">=</span>TRANSFORM</span>
<span id="cb7-98"><a href="#cb7-98" aria-hidden="true" tabindex="-1"></a>        Transformation method to be used when converting read counts [def<span class="sc">:</span> CPM]</span>
<span id="cb7-99"><a href="#cb7-99" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-100"><a href="#cb7-100" aria-hidden="true" tabindex="-1"></a><span class="sc">--</span>umccrise<span class="ot">=</span>UMCCRISE</span>
<span id="cb7-101"><a href="#cb7-101" aria-hidden="true" tabindex="-1"></a>        Directory path of the corresponding WGS<span class="sc">-</span>related data</span>
<span id="cb7-102"><a href="#cb7-102" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-103"><a href="#cb7-103" aria-hidden="true" tabindex="-1"></a><span class="sc">--</span>help, <span class="sc">-</span>h</span>
<span id="cb7-104"><a href="#cb7-104" aria-hidden="true" tabindex="-1"></a>        Show this help message and exit</span></code></pre></div>
<p><strong>Note</strong></p>
<p>Human reference genome <strong><em><a href="https://www.ncbi.nlm.nih.gov/assembly/GCF_000001405.39" class="external-link">GRCh38</a></em></strong> (<em>Ensembl</em> based annotation version <strong><em>86</em></strong>) is used for gene annotation by default. GRCh37 is no longer supported.</p>
<div class="section level3">
<h3 id="examples">Examples<a class="anchor" aria-label="anchor" href="#examples"></a>
</h3>
<p>Below are <code>RNAsum</code> CLI commands for generating HTML reports under different data availability scenarios:</p>
<ol style="list-style-type: decimal">
<li><a href="#id_1-wts-data-only">WTS data only</a></li>
<li><a href="#id_2-wts-and-wgs-data">WTS and WGS data</a></li>
<li><a href="#id_3-wts-wgs-and-clinical-data">WTS WGS and clinical data</a></li>
</ol>
<p><strong>Note</strong></p>
<ul>
<li>Example data is provided in the <code>/inst/rawdata/test_data</code> folder of the GitHub <a href="https://github.com/umccr/RNAsum" class="external-link">repo</a>.</li>
<li>The <code>RNAsum</code> runtime should be less than <strong>20 minutes</strong> using <strong>16GB RAM</strong> memory and <strong>1 CPU</strong>.</li>
</ul>
<div class="section level4">
<h4 id="id_1-wts-and-wgs-data">1. WTS and WGS data<a class="anchor" aria-label="anchor" href="#id_1-wts-and-wgs-data"></a>
</h4>
<p>This is the <strong>most frequent and preferred case</strong>, in which the <a href="#wgs">WGS</a>-based findings will be used as a primary source for expression profile prioritisation. The genome-based results can be incorporated into the report by specifying the location of the corresponding <code>umccrise</code> output files (including results from <code>PCGR</code>, <code>PURPLE</code>, and <code>Manta</code>) using the <code>--umccrise</code> argument. The <strong><code>Mutated genes</code></strong>, <strong><code>Structural variants</code></strong> and <strong><code>CN altered genes</code></strong> report sections will contain information about expression levels of the mutated genes, genes located within detected SVs and CN altered regions, respectively. The results in the <strong><code>Fusion genes</code></strong> section will be ordered based on the evidence from genome-based data. A subset of the TCGA pancreatic adenocarcinoma dataset is used as reference cohort (<code>--dataset TEST</code>).</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="ex">rnasum.R</span> <span class="dt">\</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">--sample_name</span> test_sample_WTS <span class="dt">\</span></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">--dataset</span> TEST <span class="dt">\</span></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">--dragen_rnaseq</span> inst/rawdata/test_data/dragen <span class="dt">\</span></span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">--report_dir</span> inst/rawdata/test_data/dragen/RNAsum <span class="dt">\</span></span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">--umccrise</span> inst/rawdata/test_data/umccrised/test_sample_WGS <span class="dt">\</span></span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">--save_tables</span> FALSE</span></code></pre></div>
<p>The HTML report <code>test_sample_WTS.RNAsum.html</code> will be created in the <code>inst/rawdata/test_data/dragen/RNAsum</code> folder.</p>
</div>
<div class="section level4">
<h4 id="id_2-wts-data-only">2. WTS data only<a class="anchor" aria-label="anchor" href="#id_2-wts-data-only"></a>
</h4>
<p>In this scenario, only <a href="#wts">WTS</a> data will be used and only expression levels of key <strong><a href="https://github.com/umccr/umccrise/blob/master/workflow.md#key-cancer-genes" class="external-link"><code>UMCCR Cancer genes</code></a></strong>, <strong><code>Fusion genes</code></strong>, <strong><code>Immune markers</code></strong> and homologous recombination deficiency genes (<strong><code>HRD genes</code></strong>) will be reported. Moreover, gene fusions reported in the <code>Fusion genes</code> report section will not contain information about evidence from genome-based data. A subset of the TCGA pancreatic adenocarcinoma dataset is used as the reference cohort (<code>--dataset TEST</code>).</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="ex">rnasum.R</span> <span class="dt">\</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">--sample_name</span> test_sample_WTS <span class="dt">\</span></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">--dataset</span> TEST <span class="dt">\</span></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">--dragen_rnaseq</span> inst/rawdata/test_data/dragen <span class="dt">\</span></span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">--report_dir</span> inst/rawdata/test_data/dragen/RNAsum <span class="dt">\</span></span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">--save_tables</span> FALSE</span></code></pre></div>
<p>The output HTML report <code>test_sample_WTS.RNAsum.html</code> will be created in the <code>inst/rawdata/test_data/dragen/RNAsum</code> folder.</p>
</div>
<div class="section level4">
<h4 id="id_3-wts-wgs-and-clinical-data">3. WTS WGS and clinical data<a class="anchor" aria-label="anchor" href="#id_3-wts-wgs-and-clinical-data"></a>
</h4>
<p>For samples derived from subjects, for which clinical information is available, a treatment regimen timeline can be added to the HTML report. This can be added by specifying the location of a relevant excel spreadsheet (see example <code>test_clinical_data.xlsx</code> under <code>inst/rawdata/test_data/test_clinical_data.xlsx</code>) using the <code>--clinical_info</code> argument. In this spreadsheet, at least one of the following columns is expected: <code>NEOADJUVANT REGIMEN</code>, <code>ADJUVANT REGIMEN</code>, <code>FIRST LINE REGIMEN</code>, <code>SECOND LINE REGIMEN</code> or <code>THIRD LINE REGIMEN</code>, along with <code>START</code> and <code>STOP</code> dates of corresponding treatments. A subset of the TCGA pancreatic adenocarcinoma dataset is used as the reference cohort (<code>--dataset TEST</code>).</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="ex">rnasum.R</span> <span class="dt">\</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">--sample_name</span> test_sample_WTS <span class="dt">\</span></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">--dataset</span> TEST <span class="dt">\</span></span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">--dragen_rnaseq</span> <span class="va">$(</span><span class="bu">pwd</span><span class="va">)</span>/../rawdata/test_data/dragen <span class="dt">\</span></span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">--report_dir</span> <span class="va">$(</span><span class="bu">pwd</span><span class="va">)</span>/../rawdata/test_data/dragen/RNAsum <span class="dt">\</span></span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">--umccrise</span> <span class="va">$(</span><span class="bu">pwd</span><span class="va">)</span>/../rawdata/test_data/umccrised/test_sample_WGS <span class="dt">\</span></span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">--save_tables</span> FALSE <span class="dt">\</span></span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">--clinical_info</span> <span class="va">$(</span><span class="bu">pwd</span><span class="va">)</span>/../rawdata/test_data/test_clinical_data.xlsx <span class="dt">\</span></span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">--save_tables</span> FALSE</span></code></pre></div>
<p>The HTML report <code>test_sample_WTS.RNAsum.html</code> will be created in the <code>../rawdata/test_data/stratus/test_sample_WTS_dragen_v3.9.3/RNAsum</code> folder.</p>
</div>
</div>
<div class="section level3">
<h3 id="output">Output<a class="anchor" aria-label="anchor" href="#output"></a>
</h3>
<p>The pipeline generates a HTML <strong><em>Patient Transcriptome Summary</em></strong> <strong><a href="#report">report</a></strong> and a <a href="#results">results</a> folder:</p>
<pre class="text"><code>|
|____&lt;output&gt;
  |____&lt;SampleName&gt;.&lt;output&gt;.html
  |____results
    |____exprTables
    |____glanceExprPlots
    |____...</code></pre>
<div class="section level4">
<h4 id="report">Report<a class="anchor" aria-label="anchor" href="#report"></a>
</h4>
<p>The generated HTML report includes searchable tables and interactive plots presenting expression levels of altered genes, as well as links to public resources describing the genes of interest. The report consists of several sections, including:</p>
<ul>
<li>Input data</li>
<li>Clinical information*</li>
<li>Findings summary</li>
<li>Mutated genes**</li>
<li>Fusion genes</li>
<li>Structural variants**</li>
<li>CN altered genes**</li>
<li>Immune markers</li>
<li>HRD genes</li>
<li>Cancer genes</li>
<li>Drug matching</li>
</ul>
<p>* if clinical information is available; see <code>--clinical_info</code> argument <br> ** if genome-based results are available; see <code>--umccrise</code> argument</p>
<p>Detailed description of the report structure, including result prioritisation and visualisation is available <a href="report_structure.html">here</a>.</p>
</div>
<div class="section level4">
<h4 id="results">Results<a class="anchor" aria-label="anchor" href="#results"></a>
</h4>
<p>The <code>results</code> folder contains intermediate files, including plots and tables that are presented in the HTML report.</p>
</div>
</div>
</div>
</div>

  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/umccr/RNAsum/" class="external-link">Browse source code</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing RNAsum</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Sehrish Kanwal <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0002-5044-4692" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Jacek Marzec <br><small class="roles"> Author </small>  </li>
</ul>
</div>



  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Sehrish Kanwal, Jacek Marzec.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
