<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="RNAsum">
<title>RNAsum Sections • RNAsum</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="RNAsum Sections">
<meta property="og:description" content="RNAsum">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">RNAsum</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.0.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/CHANGELOG.html">Changelog</a>
    <a class="dropdown-item" href="../articles/development.html">Development</a>
    <a class="dropdown-item" href="../articles/report_structure.html">RNAsum Sections</a>
    <a class="dropdown-item" href="../articles/tcga_projects_summary.html">TCGA Projects</a>
    <a class="dropdown-item" href="../articles/workflow.html">Workflow Details</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/umccr/RNAsum/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>RNAsum Sections</h1>
            
            <h4 data-toc-skip class="date">2024-12-17</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/umccr/RNAsum/blob/HEAD/vignettes/report_structure.Rmd" class="external-link"><code>vignettes/report_structure.Rmd</code></a></small>
      <div class="d-none name"><code>report_structure.Rmd</code></div>
    </div>

    
    
<p>The <strong><code>Mutated genes</code></strong>,
<strong><code>Structural variants</code></strong> and
<strong><code>CN altered genes</code></strong> sections will contain
information about expression levels of the mutated genes, genes located
within detected structural variants (SVs) and copy-number (CN) altered
regions, respectively. Genes will be ordered by increasing
<em>variants</em> <code>TIER</code>, <em>SV</em> <code>score</code> and
<code>CN</code> <em>value</em>, respectively, and then by decreasing
absolute values in the <code>Patient</code> vs selected
<code>dataset</code> column. Moreover, gene fusions detected in WTS data
and reported in <strong><code>Fusion genes</code></strong> section will
be first ordered based on the evidence from genome-based data
(<code>DNA support (gene A/B)</code> columns).</p>
<div class="section level3">
<h3 id="input-data">Input data<a class="anchor" aria-label="anchor" href="#input-data"></a>
</h3>
<p>Summary of the input data</p>
</div>
<div class="section level3">
<h3 id="clinical-information">Clinical information<a class="anchor" aria-label="anchor" href="#clinical-information"></a>
</h3>
<p>Treatment regimen information for patient for which clinical
information is available.</p>
<p>NOTE: for confidentiality reasons, the timeline (x-axis) projecting
patient’s treatment regimens (y-axis) is set to start from 1st January
2000, but the treatments lengths are preserved.</p>
</div>
<div class="section level3">
<h3 id="findings-summary">Findings summary<a class="anchor" aria-label="anchor" href="#findings-summary"></a>
</h3>
<p>Plot and table summarising altered genes listed across various report
sections</p>
</div>
<div class="section level3">
<h3 id="mutated-genes">Mutated genes<a class="anchor" aria-label="anchor" href="#mutated-genes"></a>
</h3>
<p>mRNA expression levels of mutated genes (containing single nucleotide
variants (SNVs) or insertions/deletions (indels)) measured in patient’s
sample and their average mRNA expression in samples from cancer patients
(from <a href="https://portal.gdc.cancer.gov/" class="external-link">TCGA</a>). This section
is available only for samples with available <a href="https://github.com/umccr/umccrise" class="external-link">umccrise</a> results.</p>
</div>
<div class="section level3">
<h3 id="fusion-genes">Fusion genes<a class="anchor" aria-label="anchor" href="#fusion-genes"></a>
</h3>
<p>Prioritised fusion genes based on <a href="https://arriba.readthedocs.io/en/latest/" class="external-link">arriba</a> results and
annotated with <a href="https://ccsm.uth.edu/FusionGDB" class="external-link">FusionGDB</a>
database. If WGS results from <a href="https://github.com/umccr/umccrise" class="external-link">umccrise</a> are available then
fusion genes in the <code>Fusion genes</code> report section are ordered
based on the evidence from genome-based data. More information about
gene fusions and methods for their detection and visualisation can be
found in the <code>fusions</code> dir.</p>
<div class="section level4">
<h4 id="prioritisation">Prioritisation<a class="anchor" aria-label="anchor" href="#prioritisation"></a>
</h4>
<p>Fusion genes detected in transcriptome data are prioritised based on
criteria ranked in the following order:</p>
<ol style="list-style-type: decimal">
<li>Involvement of fusion gene(s) <strong>detected in genomic
data</strong> (if <a href="#structural-variants">Structural variants</a>
results are available)</li>
<li>
<strong>Detected in transcriptome data</strong> by <a href="https://arriba.readthedocs.io/en/latest/" class="external-link">arriba</a>
</li>
<li>
<strong>Reported</strong> fusion event according to <a href="https://ccsm.uth.edu/FusionGDB/" class="external-link">FusionGDB</a> database</li>
<li>Decreasing number of <strong>split reads</strong>
</li>
<li>Decreasing number of <strong>pair reads</strong>
</li>
<li>Involvement of <strong>cancer gene(s)</strong> (see <a href="#cancer-genes">Cancer genes</a> section)</li>
</ol>
</div>
<div class="section level4">
<h4 id="filtering">Filtering<a class="anchor" aria-label="anchor" href="#filtering"></a>
</h4>
<p>Fusion genes detected in transcriptome data are reported if
<strong>at least one</strong> of the following criteria is met:</p>
<ol style="list-style-type: decimal">
<li>Involvement of fusion gene(s) <strong>detected in genomic
data</strong> (if SV results are available)</li>
<li>
<strong>Reported</strong> fusion event according to <a href="https://ccsm.uth.edu/FusionGDB" class="external-link">FusionGDB</a> database</li>
<li>Involvement of <strong>cancer gene(s)</strong> (see <a href="#cancer-genes">Cancer genes</a> section)</li>
<li>
<strong>Split reads</strong> &gt; 1</li>
<li>
<strong>Pair reads</strong> &gt; 1 and <strong>split reads</strong>
&gt; 1</li>
</ol>
</div>
</div>
<div class="section level3">
<h3 id="structural-variants">Structural variants<a class="anchor" aria-label="anchor" href="#structural-variants"></a>
</h3>
<p>Similar to <em>Mutated genes</em> analysis but limited to genes
located within SVs detected by <a href="https://github.com/Illumina/manta" class="external-link">Manta</a> using genomic data.
This section is available only for samples with available Manta
results.</p>
</div>
<div class="section level3">
<h3 id="cn-altered-genes">CN altered genes<a class="anchor" aria-label="anchor" href="#cn-altered-genes"></a>
</h3>
<p>Section overlaying the mRNA expression data for cancer genes with
per-gene somatic copy-number (CN) data (from PURPLE) and mutation
status, if available.</p>
</div>
<div class="section level3">
<h3 id="immune-markers">Immune markers<a class="anchor" aria-label="anchor" href="#immune-markers"></a>
</h3>
<p>Similar to <em>Mutated genes</em> analysis but limited to genes
considered to be immune markers. The immune markers used in the report
are listed in PanelApp panel <a href="https://panelapp.agha.umccr.org/panels/243/" class="external-link">Immune markers for
WTS report</a>.</p>
</div>
<div class="section level3">
<h3 id="hrd-genes">HRD genes<a class="anchor" aria-label="anchor" href="#hrd-genes"></a>
</h3>
<p>Similar to <em>Mutated genes</em> analysis but limited to genes
considered to be homologous recombination deficiency (HRD) genes. The
HRD genes used in the report are listed in PanelApp panel <a href="https://panelapp.agha.umccr.org/panels/242/" class="external-link">Homologous
recombination deficiency (HDR) for WTS report</a>.</p>
</div>
<div class="section level3">
<h3 id="cancer-genes">Cancer genes<a class="anchor" aria-label="anchor" href="#cancer-genes"></a>
</h3>
<p>Similar to analysis above, but limited to <em>UMCCR cancer
genes</em>.</p>
</div>
<div class="section level3">
<h3 id="drug-matching">Drug matching<a class="anchor" aria-label="anchor" href="#drug-matching"></a>
</h3>
<p>List of drugs targeting variants in detected <em>mutated genes</em>,
<em>fusion genes</em>, <em>structural variants-affected genes</em>,
<em>CN altered genes</em>, <em>HRD genes</em> and dysregulated
<em>cancer genes</em>, which can be considered in the treatment decision
making process.</p>
<p><strong>Note</strong></p>
<p>This section is not displayed as default. Set the
<code>--drugs</code> argument to <code>TRUE</code> to present it in the
report.</p>
</div>
<div class="section level3">
<h3 id="addendum">Addendum<a class="anchor" aria-label="anchor" href="#addendum"></a>
</h3>
<p>Additional information, including <code>Parameters</code>,
<code>Reporter details</code> and R <code>Session information</code>,
added at the end of the report.</p>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Sehrish Kanwal, Jacek Marzec.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
