<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="RNAsum">
<title>Workflow Details • RNAsum</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Workflow Details">
<meta property="og:description" content="RNAsum">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">RNAsum</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.4.14</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/development.html">Development</a>
    <a class="dropdown-item" href="../articles/report_structure.html">RNAsum Sections</a>
    <a class="dropdown-item" href="../articles/tcga_projects_summary.html">TCGA Projects</a>
    <a class="dropdown-item" href="../articles/workflow.html">Workflow Details</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/umccr/RNAsum/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Workflow Details</h1>
            
            <h4 data-toc-skip class="date">2023-07-17</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/umccr/RNAsum/blob/HEAD/vignettes/workflow.Rmd" class="external-link"><code>vignettes/workflow.Rmd</code></a></small>
      <div class="d-none name"><code>workflow.Rmd</code></div>
    </div>

    
    
<p>The <code>RNAsum</code> workflow is comprised of five main steps
shown below.</p>
<p><img src="../reference/figures/RNAsum_workflow.png" width="100%"></p>
<div class="section level2">
<h2 id="data-collection">1. Data collection<a class="anchor" aria-label="anchor" href="#data-collection"></a>
</h2>
<p>Read counts data from the patient are collected from the <a href="https://sapac.illumina.com/products/by-type/informatics-products/basespace-sequence-hub/apps/edico-genome-inc-dragen-rna-pipeline.html" class="external-link">DRAGEN
RNA</a> or <a href="https://bcbio-nextgen.readthedocs.io/en/latest/contents/bulk_rnaseq.html" class="external-link">bcbio-nextgen
RNA-seq</a> pipeline.</p>
</div>
<div class="section level2">
<h2 id="data-processing">2. Data processing<a class="anchor" aria-label="anchor" href="#data-processing"></a>
</h2>
<div class="section level3">
<h3 id="counts-processing">Counts processing<a class="anchor" aria-label="anchor" href="#counts-processing"></a>
</h3>
<p>The <strong>read count</strong> data (e.g. <a href="/inst/rawdata/test_data/final/test_sample_WTS/kallisto/abundance.tsv">abundance.tsv</a>
or <a href="/inst/rawdata/test_data/dragen/TEST.quant.sf">quant.sf</a>
quantification files from <a href="https://pachterlab.github.io/kallisto/about" class="external-link">kallisto</a> or <a href="https://salmon.readthedocs.io/en/latest/salmon.html" class="external-link">salmon</a>,
respectively) are processed according to the steps illustrated and
described below.</p>
<p><strong>Figure 1</strong></p>
<p>Counts processing scheme.</p>
<p><img src="img/counts_post-processing_scheme.png" width="40%"></p>
<div class="section level4">
<h4 id="data-collection---fig1a">Data collection - Fig1A<a class="anchor" aria-label="anchor" href="#data-collection---fig1a"></a>
</h4>
<ul>
<li>Load read count files from the following three sets of data:
<ol style="list-style-type: decimal">
<li>patient <strong>sample</strong>
</li>
<li>
<strong>external reference</strong> cohort</li>
<li>UMCCR <strong>internal reference</strong> set of in-house pancreatic
cancer samples</li>
</ol>
</li>
</ul>
</div>
<div class="section level4">
<h4 id="transformation---fig1b">Transformation - Fig1B<a class="anchor" aria-label="anchor" href="#transformation---fig1b"></a>
</h4>
<ul>
<li>Subset datasets to include common genes</li>
<li>Combine patient <strong>sample</strong> and <strong>internal
reference</strong> dataset</li>
<li>Convert counts to <strong><a href="https://haroldpimentel.wordpress.com/2014/05/08/what-the-fpkm-a-review-rna-seq-expression-units/" class="external-link">CPM</a></strong>
(<em>Counts Per Million</em>; default) or <strong><a href="https://haroldpimentel.wordpress.com/2014/05/08/what-the-fpkm-a-review-rna-seq-expression-units/" class="external-link">TPM</a></strong>
(<em>Transcripts Per Kilobase Million</em>) values in:
<ol style="list-style-type: decimal">
<li>
<strong>sample</strong> + <strong>internal reference</strong>
set</li>
<li>
<strong>external reference</strong> set</li>
</ol>
</li>
</ul>
</div>
<div class="section level4">
<h4 id="filtering-optional---fig1c">Filtering (optional) - Fig1C<a class="anchor" aria-label="anchor" href="#filtering-optional---fig1c"></a>
</h4>
<ul>
<li>Filter out genes with low counts (CPM or TPM <strong>&lt; 1</strong>
in more than 90% of samples) in:
<ol style="list-style-type: decimal">
<li>
<strong>sample</strong> + <strong>internal reference</strong>
set</li>
<li>
<strong>external reference</strong> set</li>
</ol>
</li>
</ul>
</div>
<div class="section level4">
<h4 id="normalisation-optional---fig1d">Normalisation (optional) - Fig1D<a class="anchor" aria-label="anchor" href="#normalisation-optional---fig1d"></a>
</h4>
<ul>
<li>Normalise data for sample-specific effects in:
<ol style="list-style-type: decimal">
<li>
<strong>sample</strong> + <strong>internal reference</strong>
set</li>
<li>
<strong>external reference</strong> set</li>
</ol>
</li>
</ul>
</div>
<div class="section level4">
<h4 id="combination---fig1e">Combination - Fig1E<a class="anchor" aria-label="anchor" href="#combination---fig1e"></a>
</h4>
<ul>
<li>Subset datasets to include common genes</li>
<li>Combine <strong>sample</strong> + <strong>internal
reference</strong> set with <strong>external reference</strong> set</li>
</ul>
</div>
<div class="section level4">
<h4 id="batch-effects-correction-optional---fig1f">Batch-effects correction (optional) - Fig1F<a class="anchor" aria-label="anchor" href="#batch-effects-correction-optional---fig1f"></a>
</h4>
<ul>
<li>Consider the patient <strong>sample</strong> + <strong>internal
reference</strong> (regardless of the patient sample origin) as one
batch (both sets processed with the same pipeline) and corresponding
<strong>TCGA dataset</strong> as another batch. The objective is to
remove data variation due to technical factors.</li>
</ul>
</div>
<div class="section level4">
<h4 id="data-scaling">Data scaling<a class="anchor" aria-label="anchor" href="#data-scaling"></a>
</h4>
<p>The processed count data is scaled to facilitate interpretation of
the expression values. The data is either scaled: - <strong><a href="#gene-wise-z-scoreztransformation">gene-wise</a></strong> (Z-score
transformation, default) or - <strong><a href="#group-wise-centering">group-wise</a></strong> (centering).</p>
<div class="section level5">
<h5 id="gene-wise">Gene-wise<a class="anchor" aria-label="anchor" href="#gene-wise"></a>
</h5>
<p>Z-scores are comparable by measuring the observations in multiples of
the standard deviation of the given sample. The gene-wise Z-score
transformation procedure is illustrated and described below.</p>
<p><strong>Figure 2</strong></p>
<p>Gene-wise Z-score transformation scheme.</p>
<p><img src="img/Z-score_transformation_gene_wise.png" width="30%"></p>
<ul>
<li>Extract expression values across all samples for a given
<strong>gene</strong> (<strong>Fig2A</strong>)</li>
<li>Compute <strong>Z-scores</strong> for individual samples (see
equation in <strong>Fig2B</strong>)</li>
<li>Compute <strong>median Z-scores</strong> (<strong>Fig2C</strong>)
for:
<ol style="list-style-type: decimal">
<li>
<strong>internal reference</strong> set*</li>
<li>
<strong>external reference</strong> set</li>
</ol>
</li>
<li>Present patient sample <strong>Z-score</strong> in the context of
the reference cohorts’ <strong>median Z-scores</strong>
(<strong>Fig2D</strong>).</li>
</ul>
<p>* used only for pancreatic cancer patients</p>
</div>
<div class="section level5">
<h5 id="group-wise">Group-wise<a class="anchor" aria-label="anchor" href="#group-wise"></a>
</h5>
<p>The group-wise centering apporach is illustrated and described
below.</p>
<p><strong>Figure 3</strong></p>
<p>Group-wise centering scheme.</p>
<p><img src="img/centering_group_wise.png" width="30%"></p>
<ul>
<li>Extract expression values (<strong>Fig3A</strong>) for:
<ol style="list-style-type: decimal">
<li>patient <strong>sample</strong>
</li>
<li>
<strong>internal reference</strong> set*</li>
<li>
<strong>external reference</strong> set</li>
</ol>
</li>
<li>For each gene compute <strong>median expression</strong> value
(<strong>Fig3B</strong>) in:
<ol style="list-style-type: decimal">
<li>
<strong>internal reference</strong> set*</li>
<li>
<strong>external reference</strong> set</li>
</ol>
</li>
<li>
<strong>Center</strong> the median expression values for each gene
in individual groups (<strong>Fig3C</strong>).</li>
<li>Present patient sample <strong>centered</strong> expression values
in the context the reference cohorts’ <strong>centered</strong> values
(<strong>Fig3D</strong>).</li>
</ul>
<p>* used only for pancreatic cancer patients</p>
</div>
</div>
</div>
</div>
<div class="section level2">
<h2 id="integration-with-wgs-based-results">3. Integration with WGS-based results<a class="anchor" aria-label="anchor" href="#integration-with-wgs-based-results"></a>
</h2>
<p>For patients with available WGS data processed using the <a href="https://github.com/umccr/umccrise" class="external-link">umccrise</a> pipeline, the
expression level information for mutated genes or genes located within
detected SVs/CNVs, as well as the genome-based findings, are
incorporated and used as primary source for expression profile
prioritisation.</p>
<div class="section level3">
<h3 id="somatic-snvs-and-small-indels">Somatic SNVs and small indels<a class="anchor" aria-label="anchor" href="#somatic-snvs-and-small-indels"></a>
</h3>
<ul>
<li>Check if the <a href="https://github.com/sigven/pcgr" class="external-link">PCGR</a>
tiered TSV output file is available
(e.g. ).</li>
<li>
<strong>Extract</strong> expression level information and
genome-based findings for genes with detected genomic variants (use
<code>--pcgr_tier</code> argument to define tier threshold value).</li>
<li>
<strong>Order genes</strong> by increasing variant tier and then by
decreasing absolute values representing difference between expression
levels in the patient sample and the corresponding reference
cohort.</li>
</ul>
</div>
<div class="section level3">
<h3 id="structural-variants">Structural variants<a class="anchor" aria-label="anchor" href="#structural-variants"></a>
</h3>
<ul>
<li>Check if the <a href="https://github.com/Illumina/manta" class="external-link">Manta</a>
annotated TSV output file is available (e.g. <a href="/inst/rawdata/test_data/umccrised/test_sample_WGS/structural/sv-prioritize-manta.tsv">example
input</a>).</li>
<li>
<strong>Extract</strong> expression level information and
genome-based findings for genes located within detected SVs</li>
<li>
<strong>Order genes</strong> by increasing <a href="https://github.com/vladsaveliev/simple_sv_annotation" class="external-link">SV score</a>
and then by decreasing absolute values representing difference between
expression levels in the patient sample and the corresponding reference
cohort.</li>
<li>
<strong>Compare</strong> gene fusions detected in WTS data (<a href="https://arriba.readthedocs.io/en/latest" class="external-link">arriba</a> and <a href="https://github.com/pmelsted/pizzly" class="external-link">pizzly</a>) and WGS data (<a href="https://github.com/Illumina/manta" class="external-link">Manta</a>)</li>
<li>
<strong>Prioritise</strong> WGS-supported gene fusions.</li>
</ul>
</div>
<div class="section level3">
<h3 id="somatic-cnvs">Somatic CNVs<a class="anchor" aria-label="anchor" href="#somatic-cnvs"></a>
</h3>
<ul>
<li>Check if the <a href="https://github.com/hartwigmedical/hmftools/tree/master/purple" class="external-link">PURPLE</a>
output file is available (e.g. <a href="/inst/rawdata/test_data/umccrised/test_sample_WGS/purple/purple.gene.cnv">example
input</a>).</li>
<li>
<strong>Extract</strong> expression level
<strong>information</strong> and genome-based findings for genes located
within detected CNVs (use <code>--cn_loss</code> and
<code>--cn_gain</code> arguments to define CN threshold values to
classify genes within lost and gained regions).</li>
<li>
<strong>Order genes</strong> by increasing (for genes within lost
regions) or decreasing (for genes within gained regions) Copy Number and
then by decreasing absolute values representing difference between
expression levels in the patient sample and the corresponding reference
cohort.</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="results-annotation">4. Results annotation<a class="anchor" aria-label="anchor" href="#results-annotation"></a>
</h2>
<p>WTS- and/or WGS-based results for the altered genes are collated with
knowledge derived from in-house resources and public databases (listed
below), to provide additional sources of evidence for their
significance, e.g. to flag variants with clinical significance or
potential druggable targets.</p>
<div class="section level3">
<h3 id="key-cancer-genes">Key cancer genes<a class="anchor" aria-label="anchor" href="#key-cancer-genes"></a>
</h3>
<p>The UMCCR <a href="https://github.com/vladsaveliev/NGS_Utils/blob/ac9c9a06031c6c9396a44f2edeec29a2aad4c57d/ngs_utils/reference_data/key_genes/make_umccr_cancer_genes.Rmd" class="external-link">cancer
gene panel</a> builds off from several sources:</p>
<ul>
<li>
<a href="http://bionlp.bcgsc.ca/cancermine/" class="external-link">Cancermine</a> with at
least 2 publications with at least 3 citations</li>
<li><a href="http://ncg.kcl.ac.uk/" class="external-link">NCG known cancer genes</a></li>
<li>Tier 1 <a href="https://cancer.sanger.ac.uk/census" class="external-link">COSMIC Cancer
Gene Census</a> (CGC)</li>
<li>
<a href="https://github.com/sigven/cacao" class="external-link">CACAO</a> hotspot genes
(curated from <a href="https://www.ncbi.nlm.nih.gov/clinvar/" class="external-link">ClinVar</a>, <a href="https://civicdb.org/" class="external-link">CiViC</a>, and <a href="https://www.cancerhotspots.org/" class="external-link">Cancer Hotspots</a>)</li>
<li>At least 2 matches in the following 5 sources and 8 clinical panels:
<ul>
<li>Cancer predisposition genes (<a href="https://github.com/sigven/cpsr" class="external-link">CPSR</a> list)</li>
<li>
<a href="https://cancer.sanger.ac.uk/census" class="external-link">COSMIC Cancer Gene
Census</a> (tier 2)</li>
<li>AstraZeneca 300 (AZ300)</li>
<li>Familial Cancer</li>
<li>
<a href="https://oncokb.org/" class="external-link">OncoKB</a> annotated</li>
<li>MSKC-IMPACT</li>
<li>MSKC-Heme</li>
<li>PMCC-CCP</li>
<li>Illumina-TS500</li>
<li>TEMPUS</li>
<li>Foundation One</li>
<li>Foundation Heme</li>
<li>Vogelstein</li>
</ul>
</li>
<li>Used for extracting expression levels of cancer genes (presented in
the <code>Cancer genes</code> report section)</li>
<li>Used to prioritise candidate fusion genes</li>
</ul>
</div>
<div class="section level3">
<h3 id="oncokb">OncoKB<a class="anchor" aria-label="anchor" href="#oncokb"></a>
</h3>
<ul>
<li>
<a href="https://oncokb.org/cancerGenes" class="external-link">OncoKB</a> gene list is
used to annotate altered genes across various sections in the report
(annotations and URL links in <code>External resources</code> column in
report <code>Summary tables</code>).</li>
</ul>
</div>
<div class="section level3">
<h3 id="vicc">VICC<a class="anchor" aria-label="anchor" href="#vicc"></a>
</h3>
<ul>
<li>
<a href="https://cancervariants.org/" class="external-link">Variant Interpretation for
Cancer Consortium</a> (VICC) knowledgebase is used to annotate altered
genes across various sections in the report (annotations and URL links
in <code>External resources</code> column in report
<code>Summary tables</code>).</li>
</ul>
</div>
<div class="section level3">
<h3 id="civic">CIViC<a class="anchor" aria-label="anchor" href="#civic"></a>
</h3>
<ul>
<li>The <a href="https://civicdb.org/" class="external-link">Clinical Interpretation of
Variants in Cancer</a> (CIViC) database is used to annotate altered
genes across various sections in the report (annotations and URL links
in <code>External resources</code> column in report
<code>Summary tables</code>).</li>
<li>Used to flag clinically actionable aberrations in the
<code>Drug matching</code> report section.</li>
</ul>
</div>
<div class="section level3">
<h3 id="cgi">CGI<a class="anchor" aria-label="anchor" href="#cgi"></a>
</h3>
<ul>
<li>The <a href="https://www.cancergenomeinterpreter.org/biomarkers" class="external-link">Cancer Genome
Interpreter</a> (CGI) database is used to flag genes known to be
involved in gene fusions and to prioritise candidate fusion genes.</li>
</ul>
</div>
<div class="section level3">
<h3 id="fusiongdb">FusionGDB<a class="anchor" aria-label="anchor" href="#fusiongdb"></a>
</h3>
<ul>
<li>
<a href="https://ccsm.uth.edu/FusionGDB/" class="external-link">FusionGDB</a> database is
used to flag genes known to be involved in gene fusions and to
prioritise candidate gene fusions.</li>
</ul>
</div>
<div class="section level3">
<h3 id="report-generation">5. Report generation<a class="anchor" aria-label="anchor" href="#report-generation"></a>
</h3>
<p>The final <code>RNAsum</code> HTML <strong><em>Patient Transcriptome
Summary</em></strong> report contains searchable tables and interactive
plots presenting expression levels of altered genes, as well as links to
public resources providing additional source of evidence for their
significance.</p>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Sehrish Kanwal, Jacek Marzec.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
